---
title: "Pleiotropy"
author: "Yon Ho Jee"
date: "2024-01-28"
output: html_document
---

# Fine-map pleitropic region (Loci to finemap)
```{r}
library(data.table)
#Load pleiotropy results
# don <- fread("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/pleiotropy_genes_SAIGE.csv")
don <- fread("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/pleiotropy_genes_SAIGE_May2024.csv")
don1 <- don[don$Number==max(don$Number),c(9,13:48)]
trait_list_fm <- names(don1)[don1[5,1:37]==1]

# no.genes with one or more traits
onemore<-don[don$Number>=1,] #4960
summary(don$Number)

don1 <- don[don$Number>24,c(9,13:48)]
```

# ##################
# Pleiotropy figure
# ##################
#Final pleiotropy plot
```{r}
library(ggrepel); library(ggpubr)
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HEIGHT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")

```

## Add gene & annotate most significant trait
### Gene names vertically
```{r}
axisdf <- don %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

don2 <- don %>%
  mutate(annotate_trait=ifelse(symbol %in% c("ALDH2","NAA25"),paste0(symbol,"\n","Alcohol intake"),
                                  ifelse(symbol %in% "GCKR",paste0(symbol,"\n","Triglycerides"), 
                                         ifelse(symbol %in% c("ZNF322","BTN2A1"), paste0(symbol,"\n","Height"), NA)))) 


b <- don2 %>%
    ggplot(aes(as.numeric(BPcum), Number))+ 

    # Show all points
    geom_point( aes(color=as.factor(chr)), alpha=0.8, size=3) +
    scale_color_manual(values = rep(c("#282873", "#6e90ca"), 22 )) +
    
    # custom X axis:
    #scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
    scale_x_continuous( label = c(seq(1,12),14,16,18,20), breaks= axisdf$center[axisdf$chr %in% c(seq(1,12),14,16,18,20)] ) +
    scale_y_continuous(limits = c(1,length(trait_list)), breaks=seq(0,length(trait_list),by=5) ) +     # remove space between plot area and x axis

    # Add highlighted points
    #geom_point(data=subset(don, is_highlight=="yes"), color="orange", size=2) +
  
    # Add label using ggrepel to avoid overlapping
    #geom_label_repel( data=subset(don2, is_annotate=="yes"), aes(label=symbol), size=5, max.overlaps = 10) +
    #geom_label_repel( data=subset(don2, is_annotate=="yes"), aes(label=annotate_trait), size=5, max.overlaps = 10) +
  
    geom_text_repel(
      data=subset(don2, is_annotate=="yes"), 
      #data=subset(don2, Number>20), 
      #aes(label=annotate_trait),
      aes(#as.numeric(BPcum), 20,
          label=symbol,
        segment.square  = TRUE,
        segment.inflect = TRUE,
        segment.shape   = -1),
      # force_pull   = 0, # do not pull toward data points
      force        = 0.5,
      nudge_y      = 35 - subset(don2, is_annotate=="yes")$Number,
      direction    = "x",
      angle        = 90,
      hjust        = 0,
      segment.size = 0.2,
      segment.curvature = -0.1,
      # max.iter = 1e4, max.time = 1, 
      segment.color = "red", size = 6
    ) +

    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
  xlab("Chromosome") + ylab("Number of associated traits") +
  theme(axis.text=element_text(size=14),
        axis.title=element_text(size=20), aspect.ratio = 0.7)
```

### Gene names horizontally (Revision)
```{r}
library(ggplot2)
library(ggrepel)
library(dplyr)

# Assuming don2 is already created as in your code
axisdf <- don %>% 
  group_by(chr) %>% 
  summarize(center = (max(BPcum) + min(BPcum)) / 2)

don2 <- don %>%
  mutate(annotate_trait = ifelse(symbol %in% c("ALDH2", "NAA25"), paste0(symbol, "\n", "Alcohol intake"),
                            ifelse(symbol %in% "GCKR", paste0(symbol, "\n", "Triglycerides"), 
                            ifelse(symbol %in% c("ZNF322", "BTN2A1"), paste0(symbol, "\n", "Height"), NA))))

# Plot
b <- don2 %>%
  ggplot(aes(as.numeric(BPcum), Number)) + 

  # Show all points
  geom_point(aes(color = as.factor(chr)), alpha = 0.8, size = 3) +
  scale_color_manual(values = rep(c("#282873", "#6e90ca"), 22)) +

  # Custom X axis
  scale_x_continuous(labels = c(seq(1, 12), 14, 16, 18, 20), 
                     breaks = axisdf$center[axisdf$chr %in% c(seq(1, 12), 14, 16, 18, 20)]) +
  scale_y_continuous(limits = c(1, length(trait_list)), breaks = seq(0, length(trait_list), by = 5)) +

  # Add label using geom_text_repel to add horizontal labels near the data points
  geom_text_repel(
    data = subset(don2, is_annotate == "yes"), 
    aes(label = symbol), 
    nudge_y = 5,   # Adjust vertical distance from the points
    nudge_x = 50000, # Adjust horizontal distance from the points
    direction = "both",  # Ensure labels move vertically, not horizontally
    hjust = 0,       # Left-align the labels
    size = 5,        # Text size
    force = 0.5,     # Adjust repelling force
    max.overlaps = 20, # Limit overlapping labels
    segment.curvature = -0.1,
    arrow = arrow(length = unit(0.015, "npc"), type = "closed"),  # Arrows pointing to points
    segment.size = 0.3,   # Size of the arrow segment
    segment.color = "red" # Color of the arrow segment
  ) +

  # Custom the theme
  theme_bw() +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text = element_text(size = 14),
    axis.title = element_text(size = 20),
    aspect.ratio = 0.7
  ) +
  xlab("Chromosome") + 
  ylab("Number of associated traits")

# Display the plot
b
```

```{r}
b1 <- don2 %>%
    ggplot(aes(as.numeric(BPcum), Number))+ 

    # Show all points
    geom_point( aes(color=as.factor(chr)), alpha=0.8, size=3) +
    scale_color_manual(values = rep(c("#282873", "#6e90ca"), 22 )) +
    
    # custom X axis:
    #scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
    scale_x_continuous( label = c(seq(1,12),14,16,18,20), breaks= axisdf$center[axisdf$chr %in% c(seq(1,12),14,16,18,20)] ) +
    scale_y_continuous(limits = c(1,length(trait_list)), breaks=seq(0,length(trait_list),by=5) ) +     # remove space between plot area and x axis

    # Add highlighted points
    #geom_point(data=subset(don, is_highlight=="yes"), color="orange", size=2) +
  
    # Add label using ggrepel to avoid overlapping
    #geom_label_repel( data=subset(don2, is_annotate=="yes"), aes(label=symbol), size=5, max.overlaps = 10) +
    #geom_label_repel( data=subset(don2, is_annotate=="yes"), aes(label=annotate_trait), size=5, max.overlaps = 10) +
  
    geom_text_repel(
      data=subset(don2, is_annotate=="yes"), 
      #data=subset(don2, Number>20), 
      #aes(label=annotate_trait),
      aes(as.numeric(BPcum), 27,
          label=symbol,
        segment.square  = TRUE,
        segment.inflect = TRUE,
        segment.shape   = -1),
      # force_pull   = 0, # do not pull toward data points
      force        = 0.5,
      nudge_y      = 3,
      direction    = "x",
      angle        = 90,
      hjust        = 0,
      segment.size = 0.2,
      segment.curvature = -0.1,
      # max.iter = 1e4, max.time = 1, 
      segment.color = "red"
    ) +

    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    ) +
  xlab("Chromosome") + ylab("Number of associated traits") +
  theme(axis.text=element_text(size=20),
        axis.title=element_text(size=20,face="bold"), aspect.ratio = 0.7)
```

# ##############################
# Number of significant loci
# ##############################
#3/6/2024
##added fraction novel
```{r}
library(dplyr);library(reshape2); library(ggplot2);library(tidyr)

#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/yjee/FUMA/SAIGE/loci_count.csv")
#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg.csv")
#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg_gwcat.csv")
res <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg_gwcat_HEIGHT_TSH_EAS_500kb_10052024.csv",data.table = F)

res1 <- res %>%
  mutate(N_rep = Number.of.loci - Number.of.novel.loci) %>%
  rename(Novel=Number.of.novel.loci, Known=N_rep)

res1 <- res1%>% 
       mutate(category = case_when(
         (trait %in% c("BMI","HEIGHT","WAIST","WT")) ~ "Anthropometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardio- vascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid", #"Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor", #"Tumor marker",
              TRUE ~ ""))
  
res1$category <- factor(res1$category,
                        levels=c(
                          "Anthropometry","Metabolic","Hematological","Cardio- vascular","Kidney","Liver",
                          #"Thyroid hormone","Tumor marker",
                          "Thyroid","Tumor",
                          "Lifestyle"))

res1$description <- c("Low density lipoprotein","Triglyceride","High density lipoprotein","Total cholestrol","Fasting blood sugar","Insulin","Glutamic oxaloacetic transaminase","Glutamic pyruvic transaminase","Gamma-glutamyl Transferase","Total bilirubin","Albumin","Thyroid-stimulating hormone","Carcinoembryonic antigen","Creatinine","Adiponectin","Uric acid","Alkaline phosphatase","Hemoglbin","Hematocrit","Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Red blood cell","Red blood cell distribution width","White blood cell","Platelet","Eosinophil","Systolic blood pressure", "Diastolic blood pressure","Weight","Height","Body mass index","Waist circumference","Smoking amount","Alcohol intake","Coffee intake")

res1$trait <- factor(res1$trait, levels = c("ALCO_AMOUNT","COFFA","SMOKA_MOD",
                                            "CEA",
                                            "TSH",
                                            "ALB","GGT","GOT","GPT","BIL",
                                            "ALP","CREAT","URIC",
                                            "DBP","SBP",
                                            "EOS","HCT","HB","MCH","MCHC","MCV","PLT","RBC","RDW","WBC",
                                            "ADIPO","FBS","HDL","INSULIN","LDL","CHO","TG",
                                            "BMI","HEIGHT","WAIST","WT"))
res1$description <- factor(res1$description, levels = c("Alcohol intake","Coffee intake","Smoking amount",
                                            "Carcinoembryonic antigen",
                                            "Thyroid-stimulating hormone",
                                            "Albumin","Gamma-glutamyl Transferase","Glutamic oxaloacetic transaminase","Glutamic pyruvic transaminase","Total bilirubin",
                                            "Alkaline phosphatase","Creatinine","Uric acid",
                                            "Diastolic blood pressure","Systolic blood pressure",
                                            "Eosinophil","Hematocrit","Hemoglbin","Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Platelet","Red blood cell","Red blood cell distribution width","White blood cell",
                                            "Adiponectin","Fasting blood sugar","High density lipoprotein","Insulin","Low density lipoprotein","Total cholestrol","Triglyceride",
                                            "Body mass index","Height","Waist circumference","Weight"))

res1$description <- factor(res1$description, levels = c("Alcohol intake","Smoking amount","Coffee intake",
                                            "Carcinoembryonic antigen",
                                            
                                            "Thyroid-stimulating hormone",
                                            
                                            "Gamma-glutamyl Transferase","Glutamic pyruvic transaminase","Albumin","Total bilirubin","Glutamic oxaloacetic transaminase",
                                            
                                            "Alkaline phosphatase","Creatinine","Uric acid",
                                            
                                            "Diastolic blood pressure","Systolic blood pressure",
                                            
                                            "Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Eosinophil","Platelet","Red blood cell distribution width","White blood cell","Red blood cell","Hematocrit","Hemoglbin",
                                            
                                            "High density lipoprotein","Triglyceride","Total cholestrol","Low density lipoprotein","Fasting blood sugar","Adiponectin","Insulin",
                                            
                                            "Height","Body mass index","Weight","Waist circumference"))

#meta_plot_trait_long <- gather(meta_plot_trait, report, N, N_new:N_rep, factor_key=TRUE)

n <- 9
h1 <- hcl.colors(n, palette = "Dynamic")

res1 <- res1%>% 
       mutate(col_cat = case_when(
         (category %in% c("Anthropometry")) ~ h1[1],
         (category %in% c("Cardio- vascular")) ~ h1[4],
         (category %in% c("Hematological")) ~ h1[3],
         (category %in% c("Kidney")) ~ h1[5],
         (category %in% c("Lifestyle")) ~ h1[9],
         (category %in% c("Liver")) ~ h1[6],
              (category %in% c("Metabolic")) ~ h1[2],
              # (category %in% c("Thyroid hormone")) ~ h1[7],
              # (category %in% c("Tumor marker")) ~ h1[8],
              (category %in% c("Thyroid")) ~ h1[7],
              (category %in% c("Tumor")) ~ h1[8],
              TRUE ~ ""))
res1$col_cat <- factor(res1$col_cat, levels =h1)

res1$perc<-res1$Novel/res1$Number.of.loci*100
res1_long <- gather(res1, report, N, Novel:Known, factor_key=TRUE)
```

```{r}
library(ggstats)
library(ggh4x)

my.theme = list(theme(axis.line.y.left = element_line(size = 1, color = "white"),
                      axis.line.y.right = element_line(size = 1, color = "black"),
                      axis.line.x.top =  element_line(size = 1, color = "white"),
                      axis.line.x.bottom = element_line(size = 1, color = "black")))

ridiculous_strips <- strip_themed(
     # Horizontal strips
     background_x = elem_list_rect(fill = c("limegreen", "dodgerblue")),
     text_x = elem_list_text(colour = c("dodgerblue", "limegreen"),
                             face = c("bold", "bold")),
     by_layer_x = TRUE,
     # Vertical strips
     background_y = elem_list_rect(
       fill = h1
     ),
     text_y = elem_list_text(angle = c(90,90,90,360,90,90,360,360,90),
                             #angle = rep(90, 9),
                             face = c("bold","bold","bold","bold","bold","bold","bold","bold","bold"),
                             #face = rep("bold", 9),
                             size=c(11,15,15,11,15,15,11,11,14)),
     by_layer_y = FALSE
)

# my_cols <- c("steelblue", "gold", "firebrick", "forestgreen", 
#              "orange", "white", "grey", "black")
# palette(my_cols)

a1 <- ggplot(data = res1_long, aes(x = factor(description), y = N, fill = factor(report,levels = c("Known","Novel")))) + 
  geom_bar(stat="identity", aes(y=N, fill = "grey"),alpha = .5) + 
  geom_bar(data = subset(res1_long, report=="Novel"), stat="identity", fill = sort(res1_long$col_cat[res1_long$report=="Novel"]), alpha = .8) + 
  coord_flip() +  
   facet_grid2(category~., scales = "free", space = "free",switch = "x", labeller = label_wrap_gen(width = 2, multi_line = TRUE), strip = ridiculous_strips) +
    ylab("Number of significant loci") + xlab("") + 
  theme(strip.text.y = element_text(angle = 270, size = 8)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        #axis.line = element_line(colour = "black"), 
        legend.title =element_blank(), legend.text=element_text(size=20), 
        legend.position = c(0.2, 0.8),
        axis.title=element_text(size=15),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())  + my.theme +
  #scale_fill_brewer(palette = "Dark2") + 
  scale_fill_manual(labels = c("Known","Novel"), values = c("grey", "black")) +
  scale_y_reverse() +
    geom_text(data=subset(res1_long, report=="Known"), aes(label = Number.of.loci,y=Number.of.loci+5),
              position = position_dodge(width = 1), vjust= 0.3, hjust = 1) 
  #theme(strip.background = element_rect(fill = fills))

colfill <- c("#E494AF","#E494AF","#E494AF","#D596D2","#ACA4E2","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#38BDBB","#38BDBB","#38BDBB","#5CBD92","#5CBD92","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B666","#93B66E","#BFAA67","#BFAA67","#BFAA67","#BFAA67","#BFAA76","#BFAA76","#BFAA67","#DB9D85","#DB9D85","#DB9D85","#DB9D85")


ridiculous_strips2 <- strip_themed(
     # Horizontal strips
     background_x = elem_list_rect(fill = c("limegreen", "dodgerblue")),
     text_x = elem_list_text(colour = c("dodgerblue", "limegreen"),
                             face = c("bold", "bold")),
     by_layer_x = TRUE,
     # Vertical strips
     background_y = elem_list_rect(
       fill = rep("white",9)
     ),
     text_y = elem_list_text(angle = c(90,90,90,360,90,90,360,360,90),
                             #angle = rep(90, 9),
                             face = c("bold","bold","bold","bold","bold","bold","bold","bold","bold"),
                             #face = rep("bold", 9),
                             size=c(11,15,15,11,15,15,11,11,14),
                             colour = rep("white",9)),
     
     by_layer_y = FALSE
)

a2 <- ggplot(data = res1_long, aes(x = factor(description), y = perc, fill = factor(report,levels = c("Known","Novel")))) + 
  #geom_bar(stat="identity") + 
  #geom_bar(data = subset(res1_long, report=="Known"), stat="identity", fill = "grey", alpha = .5) +
  geom_bar(data = subset(res1_long, report=="Novel"), stat="identity", fill = sort(res1_long$col_cat[res1_long$report=="Novel"]), alpha = .8) + 
  
  #geom_text(data = subset(res1_long, report=="Novel"), stat = "identity", position = position_fill(.5),aes(label=perc)) + 
  geom_text(data=subset(res1_long, report=="Novel"), aes(label = paste0(round(perc,1),"%"), y=perc+10),
              position = position_dodge(width = 1), vjust= 0.3, hjust = 1) +
  #geom_text(data = subset(res1_long, report=="Novel"), stat = "prop", position = position_fill(.5)) + 
  #geom_text(stat = "prop", position = position_fill(.5)) + 
  coord_flip() +  
  facet_grid2(category~., scales = "free", space = "free",switch = "x", labeller = label_wrap_gen(width = 2, multi_line = TRUE), strip = ridiculous_strips2) +
  
  ylab("Fraction of novel loci") + xlab("") + 
  theme(strip.text.y = element_text(angle = 270, size = 40)) +
  
  #theme(strip.placement = "outside")+

  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"), 
        legend.title =element_blank(), legend.text=element_text(size=15),
        axis.text.y = element_text(size = 15,
                                   #color = unique(colfill),
                                   face =  "bold"),
        axis.title=element_text(size=15),
        legend.position = "none")  
  

library(grid)
library(gridExtra)
gg1 <- ggplot_gtable(ggplot_build(a1))
gg2 <- ggplot_gtable(ggplot_build(a2))

grid.arrange(gg1,gg2,ncol=2,widths=c(0.45,0.55))
```

```{r}
# pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_number_novel_loci.pdf",height=12, width=15)
# pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_number_novel_loci_otg.pdf",height=12, width=15)
# pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_number_novel_loci_otg_gwcat.pdf",height=12, width=15)
# pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_number_novel_loci_otg_gwcat_v2.pdf",height=12, width=15)
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_number_novel_loci_otg_gwcat_10072024.pdf",height=12, width=15)

ggarrange(gg1,gg2, ncol = 2, widths=c(0.45,0.55),
         labels = c("a"),font.label=list(size=20))
#grid.arrange(gg1,gg2,ncol=2,widths=c(0.45,0.55))
dev.off() # Close the file
```

## ASHG 24' slides
```{r}
library(dplyr);library(reshape2); library(ggplot2);library(tidyr)

#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/yjee/FUMA/SAIGE/loci_count.csv")
#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg.csv")
#res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg_gwcat.csv")
res <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/OTG/kcps2_loci_count_otg_gwcat_HEIGHT_TSH_EAS_500kb_10052024.csv",data.table = F)

res1 <- res %>%
  mutate(N_rep = Number.of.loci - Number.of.novel.loci) %>%
  rename(Novel=Number.of.novel.loci, Known=N_rep)

res1 <- res1%>% 
       mutate(category = case_when(
         (trait %in% c("BMI","HEIGHT","WAIST","WT")) ~ "Anthro- pometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardio- vascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid", #"Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor", #"Tumor marker",
              TRUE ~ ""))
  
res1$category <- factor(res1$category,
                        levels=c(
                          "Anthro- pometry","Metabolic","Hematological","Cardio- vascular","Kidney","Liver",
                          #"Thyroid hormone","Tumor marker",
                          "Thyroid","Tumor",
                          "Lifestyle"))

res1$description <- c("Low density lipoprotein","Triglyceride","High density lipoprotein","Total cholestrol","Fasting blood sugar","Insulin","Glutamic oxaloacetic transaminase","Glutamic pyruvic transaminase","Gamma-glutamyl Transferase","Total bilirubin","Albumin","Thyroid-stimulating hormone","Carcinoembryonic antigen","Creatinine","Adiponectin","Uric acid","Alkaline phosphatase","Hemoglbin","Hematocrit","Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Red blood cell","Red blood cell distribution width","White blood cell","Platelet","Eosinophil","Systolic blood pressure", "Diastolic blood pressure","Weight","Height","Body mass index","Waist circumference","Smoking amount","Alcohol intake","Coffee intake")

res1$trait <- factor(res1$trait, levels = c("ALCO_AMOUNT","COFFA","SMOKA_MOD",
                                            "CEA",
                                            "TSH",
                                            "ALB","GGT","GOT","GPT","BIL",
                                            "ALP","CREAT","URIC",
                                            "DBP","SBP",
                                            "EOS","HCT","HB","MCH","MCHC","MCV","PLT","RBC","RDW","WBC",
                                            "ADIPO","FBS","HDL","INSULIN","LDL","CHO","TG",
                                            "BMI","HEIGHT","WAIST","WT"))
res1$description <- factor(res1$description, levels = c("Alcohol intake","Coffee intake","Smoking amount",
                                            "Carcinoembryonic antigen",
                                            "Thyroid-stimulating hormone",
                                            "Albumin","Gamma-glutamyl Transferase","Glutamic oxaloacetic transaminase","Glutamic pyruvic transaminase","Total bilirubin",
                                            "Alkaline phosphatase","Creatinine","Uric acid",
                                            "Diastolic blood pressure","Systolic blood pressure",
                                            "Eosinophil","Hematocrit","Hemoglbin","Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Platelet","Red blood cell","Red blood cell distribution width","White blood cell",
                                            "Adiponectin","Fasting blood sugar","High density lipoprotein","Insulin","Low density lipoprotein","Total cholestrol","Triglyceride",
                                            "Body mass index","Height","Waist circumference","Weight"))

res1$description <- factor(res1$description, levels = c("Alcohol intake","Smoking amount","Coffee intake",
                                            "Carcinoembryonic antigen",
                                            
                                            "Thyroid-stimulating hormone",
                                            
                                            "Gamma-glutamyl Transferase","Glutamic pyruvic transaminase","Albumin","Total bilirubin","Glutamic oxaloacetic transaminase",
                                            
                                            "Alkaline phosphatase","Creatinine","Uric acid",
                                            
                                            "Diastolic blood pressure","Systolic blood pressure",
                                            
                                            "Mean corpuscular hemoglobin","Mean corpuscular hemoglobin concentration","Mean corpuscular volume","Eosinophil","Platelet","Red blood cell distribution width","White blood cell","Red blood cell","Hematocrit","Hemoglbin",
                                            
                                            "High density lipoprotein","Triglyceride","Total cholestrol","Low density lipoprotein","Fasting blood sugar","Adiponectin","Insulin",
                                            
                                            "Height","Body mass index","Weight","Waist circumference"))

#meta_plot_trait_long <- gather(meta_plot_trait, report, N, N_new:N_rep, factor_key=TRUE)

n <- 9
h1 <- hcl.colors(n, palette = "Dynamic")

res1 <- res1%>% 
       mutate(col_cat = case_when(
         (category %in% c("Anthro- pometry")) ~ h1[1],
         (category %in% c("Cardio- vascular")) ~ h1[4],
         (category %in% c("Hematological")) ~ h1[3],
         (category %in% c("Kidney")) ~ h1[5],
         (category %in% c("Lifestyle")) ~ h1[9],
         (category %in% c("Liver")) ~ h1[6],
              (category %in% c("Metabolic")) ~ h1[2],
              # (category %in% c("Thyroid hormone")) ~ h1[7],
              # (category %in% c("Tumor marker")) ~ h1[8],
              (category %in% c("Thyroid")) ~ h1[7],
              (category %in% c("Tumor")) ~ h1[8],
              TRUE ~ ""))
res1$col_cat <- factor(res1$col_cat, levels =h1)

res1$perc<-res1$Novel/res1$Number.of.loci*100
res1_long <- gather(res1, report, N, Novel:Known, factor_key=TRUE)
```

```{r}
library(ggstats)
library(ggh4x)

my.theme = list(theme(axis.line.y.left = element_line(size = 1, color = "white"),
                      axis.line.y.right = element_line(size = 1, color = "black"),
                      axis.line.x.top =  element_line(size = 1, color = "white"),
                      axis.line.x.bottom = element_line(size = 1, color = "black")))

ridiculous_strips <- strip_themed(
     # Horizontal strips
     background_x = elem_list_rect(fill = c("limegreen", "dodgerblue")),
     text_x = elem_list_text(colour = c("dodgerblue", "limegreen"),
                             face = c("bold", "bold")),
     by_layer_x = TRUE,
     # Vertical strips
     background_y = elem_list_rect(
       fill = h1
     ),
     text_y = elem_list_text(angle = c(360,90,90,360,360,360,360,360,360),
                             #angle = rep(90, 9),
                             face = c("bold","bold","bold","bold","bold","bold","bold","bold","bold"),
                             #face = rep("bold", 9),
                             size=c(11,15,15,11,15,15,11,11,14)),
     by_layer_y = FALSE
)

# my_cols <- c("steelblue", "gold", "firebrick", "forestgreen", 
#              "orange", "white", "grey", "black")
# palette(my_cols)

a1 <- ggplot(data = res1_long, aes(x = factor(description), y = N, fill = factor(report,levels = c("Known","Novel")))) + 
  geom_bar(stat="identity", aes(y=N, fill = "grey"),alpha = .5) + 
  geom_bar(data = subset(res1_long, report=="Novel"), stat="identity", fill = sort(res1_long$col_cat[res1_long$report=="Novel"]), alpha = .8) + 
  coord_flip() +  
   facet_grid2(category~., scales = "free", space = "free",switch = "x", labeller = label_wrap_gen(width = 2, multi_line = TRUE), strip = ridiculous_strips) +
    ylab("Number of significant loci") + xlab("") + 
  theme(strip.text.y = element_text(angle = 270, size = 8)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        #axis.line = element_line(colour = "black"), 
        legend.title =element_blank(), legend.text=element_text(size=20), 
        legend.position = c(0.2, 0.8),
        axis.title=element_text(size=20),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())  + my.theme +
  #scale_fill_brewer(palette = "Dark2") + 
  scale_fill_manual(labels = c("Known","Novel"), values = c("grey", "black")) +
  scale_y_reverse() +
    geom_text(data=subset(res1_long, report=="Known"), aes(label = Number.of.loci,y=Number.of.loci+5),
              position = position_dodge(width = 1), vjust= 0.3, hjust = 1, size=6) 
  #theme(strip.background = element_rect(fill = fills))

colfill <- c("#E494AF","#E494AF","#E494AF","#D596D2","#ACA4E2","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#38BDBB","#38BDBB","#38BDBB","#5CBD92","#5CBD92","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B666","#93B66E","#BFAA67","#BFAA67","#BFAA67","#BFAA67","#BFAA76","#BFAA76","#BFAA67","#DB9D85","#DB9D85","#DB9D85","#DB9D85")


ridiculous_strips2 <- strip_themed(
     # Horizontal strips
     background_x = elem_list_rect(fill = c("limegreen", "dodgerblue")),
     text_x = elem_list_text(colour = c("dodgerblue", "limegreen"),
                             face = c("bold", "bold")),
     by_layer_x = TRUE,
     # Vertical strips
     background_y = elem_list_rect(
       fill = rep("white",9)
     ),
     text_y = elem_list_text(angle = c(90,90,90,360,90,90,360,360,90),
                             #angle = rep(90, 9),
                             face = c("bold","bold","bold","bold","bold","bold","bold","bold","bold"),
                             #face = rep("bold", 9),
                             size=c(11,15,15,11,15,15,11,11,14),
                             colour = rep("white",9)),
     
     by_layer_y = FALSE
)

a2 <- ggplot(data = res1_long, aes(x = factor(description), y = perc, fill = factor(report,levels = c("Known","Novel")))) + 
  #geom_bar(stat="identity") + 
  #geom_bar(data = subset(res1_long, report=="Known"), stat="identity", fill = "grey", alpha = .5) +
  geom_bar(data = subset(res1_long, report=="Novel"), stat="identity", fill = sort(res1_long$col_cat[res1_long$report=="Novel"]), alpha = .8) + 
  
  #geom_text(data = subset(res1_long, report=="Novel"), stat = "identity", position = position_fill(.5),aes(label=perc)) + 
  geom_text(data=subset(res1_long, report=="Novel"), aes(label = paste0(round(perc,1),"%"), y=perc+10),
              position = position_dodge(width = 1), vjust= 0.3, hjust = 1, size=6) +
  #geom_text(data = subset(res1_long, report=="Novel"), stat = "prop", position = position_fill(.5)) + 
  #geom_text(stat = "prop", position = position_fill(.5)) + 
  coord_flip() +  
  facet_grid2(category~., scales = "free", space = "free",switch = "x", labeller = label_wrap_gen(width = 2, multi_line = TRUE), strip = ridiculous_strips2) +
  
  ylab("Fraction of novel loci") + xlab("") + 
  theme(strip.text.y = element_text(angle = 270, size = 40)) +
  
  #theme(strip.placement = "outside")+

  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"), 
        legend.title =element_blank(), legend.text=element_text(size=20),
        axis.text.y = element_text(size = 15,
                                   #color = unique(colfill),
                                   face =  "bold"),
        axis.title=element_text(size=20),
        legend.position = "none")  
  

library(grid)
library(gridExtra)
gg1 <- ggplot_gtable(ggplot_build(a1))
gg2 <- ggplot_gtable(ggplot_build(a2))

grid.arrange(gg1,gg2,ncol=2,widths=c(0.45,0.55))
```

# ##############################
# rg_rp_plot
# ##############################
# rg_rp_plot.Rmd
```{r}
library("xlsx")
data3<-read.xlsx('/n/holylfs05/LABS/kraft_lab/Lab/yjee/ldsc/output/PLINK_irnt/rg_rp_cor_pval.xlsx',1)

all_rg_out_nomiss<-fread('/n/holylfs05/LABS/kraft_lab/Lab/yjee/ldsc/output/PLINK_irnt/rg_outputs_PLINK_irnt.txt', data.table = F) %>%
  select(-rg)

rg_rp_se <- left_join(all_rg_out_nomiss, data3, by=c("p1"="phenotype1","p2"="phenotype2")) %>%
  mutate(fdr_cat = case_when(
    rg_fdr<0.05 & rp_fdr<0.05 ~ "b", #"both signicant",
    rg_fdr<0.05 & rp_fdr>=0.05 ~ "g", #"rg signicant",
    rg_fdr>=0.05 & rp_fdr<0.05 ~ "p", #"rp signicant",
    rg_fdr>=0.05 & rp_fdr>=0.05 ~ "n")) #"neither signicant"))
```
```{r}
summary(rg_rp_se)
tmp <- rg_rp_se %>%
  filter(rg>0)
summary(rg_rp_se$rp[rg_rp_se$rg>0])
tmp1 <- rg_rp_se %>%
  filter(rg>0.1,rp<0.015,fdr_cat=="g") %>%
  select(p1,p2,rg,rg_fdr,rp,rp_fdr,fdr_cat)
tmp2 <- rg_rp_se %>%
  filter(fdr_cat=="g")
tmp3 <- rg_rp_se %>%
  filter(fdr_cat=="p")
tmp4 <- rg_rp_se %>%
  filter(rg*rp<0,fdr_cat!="n") %>%
  select(p1,p2,rg,rg_fdr,rp,rp_fdr,fdr_cat)
table(tmp4$fdr_cat)
 #  b   g   p 
 # 28   4 112 
tmp5 <- rg_rp_se %>%
  filter(rg*rp<0,fdr_cat=="b") %>%
  select(p1,p2,rg,rg_fdr,rp,rp_fdr,fdr_cat)

```

```{r}
#palette
#cbbPalette <- c("purple4","darkred","#0072B2","grey")
cbbPalette <- c("#a161d1","#d16163","#618dd1","grey")
#cbbPalette <- c("#512888","#B11225","#062A77","grey")

s2 <- rg_rp_se %>%
  ggplot(aes(y=rg,x=rp,colour=fdr_cat))+ 
  geom_point(position = position_dodge(0.5), size = 3)+
  #geom_pointrange(aes(xmin=rg-1.96*se, xmax=rg+1.96*se)) +
  #scale_color_continuous(trans='reverse') +
  geom_label_repel(
    aes(label=paste0(p1,"-",p2)), size=4, box.padding = unit(0.5, "lines"), max.overlaps = 30,
    show.legend = F
  )+
  geom_smooth(method=lm, formula = y~x, color="grey5") +
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype="dotted") + 
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") + 
  #scale_colour_manual(values=cbPalette)+
  scale_colour_manual(name="FDR",
                        labels = c("Both signicant",bquote("Only"~r[g]*" significant"), bquote("Only"~r[p]*" significant"), "Neither significant"), 
                        breaks = c("b", "g", "p", "n"), 
                        values = cbbPalette) +

  xlim(-0.7,1) + ylim(-0.7,1) + 
  xlab(bquote(r[p])) + ylab(bquote(r[g])) +
  theme(legend.position = "right", 
        legend.key=element_blank(), legend.text = element_text(size=15), 
        legend.title = element_text(size=15),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 20),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 20),
        panel.border = element_rect(colour = "black", fill=NA))
```

## ASHG 24' slides
Hb
```{r}
cbbPalette <- c("#a161d1","#d16163","#618dd1","grey")
traits_to_filter <- c("HB", "BMI", "FBS", "SBP", "DBP","TG")

# Create a new column to set the color for specific data points
rg_rp_se1 <- rg_rp_se %>%
  mutate(color_group = ifelse(((p1 %in% "HB" & p2 %in% traits_to_filter)|(p1 %in% traits_to_filter & p2 %in% "HB")) & fdr_cat == "b", fdr_cat, "grey90")) %>%
  mutate(p1=ifelse(p1=="HB","Hb",p1),
         p2=ifelse(p2=="HB","Hb",p2))

table(rg_rp_se1$color_group)
# Plot the data
s20 <- rg_rp_se1 %>%
  ggplot(aes(y = rg, x = rp)) +
  
  # Plot grey points first
  geom_point(data = subset(rg_rp_se1, color_group == "grey90"), aes(colour = "grey90"), position = position_dodge(0.5), size = 3, show.legend = FALSE) +
  
  # Plot purple points on top (Both significant)
  geom_point(data = subset(rg_rp_se1, color_group == "b"), aes(colour = "b"), position = position_dodge(0.5), size = 3) +
  
  # Add labels only for traits_to_filter
  geom_label_repel(
    data = subset(rg_rp_se1, (p1 %in% "Hb" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "Hb")& fdr_cat == "b"),
    aes(label = paste0(p1, "-", p2)),
    size = 4, box.padding = unit(0.5, "lines"), max.overlaps = 30,
    show.legend = FALSE
  ) +
  
  # geom_smooth(method = lm, formula = y ~ x, color = "grey5") +
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_vline(xintercept = 0, linetype = "dotted") + 
  scale_colour_manual(name = "Highlight",
                      labels = c("Both significant", "Other"), 
                      breaks = c("b", "grey90"), 
                      values = c(cbbPalette[1], "grey90")) +
  xlim(-0.7, 1) + ylim(-0.7, 1) + 
  xlab(bquote(r[p])) + ylab(bquote(r[g])) +
  theme(legend.position = "none", 
        legend.key = element_blank(), legend.text = element_text(size = 15), 
        legend.title = element_text(size = 15),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 20),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 20),
        panel.border = element_rect(colour = "black", fill = NA))

```

HDL
```{r}
cbbPalette <- c("#a161d1","#d16163","#618dd1","grey")
traits_to_filter <- c("HDL", "BMI", "FBS", "INSULIN")

# Create a new column to set the color for specific data points
rg_rp_se1 <- rg_rp_se %>%
  mutate(color_group = ifelse((p1 %in% "HDL" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "HDL") & fdr_cat == "b", fdr_cat, "grey90"))

table(rg_rp_se1$color_group)
# Plot the data
s21 <- rg_rp_se1 %>%
  ggplot(aes(y = rg, x = rp)) +
  
  # Plot grey points first
  geom_point(data = subset(rg_rp_se1, color_group == "grey90"), aes(colour = "grey90"), position = position_dodge(0.5), size = 3, show.legend = FALSE) +
  
  # Plot purple points on top (Both significant)
  geom_point(data = subset(rg_rp_se1, color_group == "b"), aes(colour = "b"), position = position_dodge(0.5), size = 3) +
  
  # Add labels only for traits_to_filter
  geom_label_repel(
    data = subset(rg_rp_se, (p1 %in% "HDL" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "HDL")& fdr_cat == "b"),
    aes(label = paste0(p1, "-", p2)),
    size = 4, box.padding = unit(0.5, "lines"), max.overlaps = 30,
    show.legend = FALSE
  ) +
  
  # geom_smooth(method = lm, formula = y ~ x, color = "grey5") +
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_vline(xintercept = 0, linetype = "dotted") + 
  scale_colour_manual(name = "Highlight",
                      labels = c("Both significant", "Other"), 
                      breaks = c("b", "grey90"), 
                      values = c(cbbPalette[1], "grey90")) +
  xlim(-0.7, 1) + ylim(-0.7, 1) + 
  xlab(bquote(r[p])) + ylab(bquote(r[g])) +
  theme(legend.position = "none", 
        legend.key = element_blank(), legend.text = element_text(size = 15), 
        legend.title = element_text(size = 15),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 20),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 20),
        panel.border = element_rect(colour = "black", fill = NA))

```

ADIPO
```{r}
cbbPalette <- c("#a161d1","#d16163","#618dd1","grey")
traits_to_filter <- c("ADIPO", "BMI", "FBS", "INSULIN")

# Create a new column to set the color for specific data points
rg_rp_se1 <- rg_rp_se %>%
  mutate(color_group = ifelse((p1 %in% "ADIPO" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "ADIPO") & fdr_cat == "b", fdr_cat, "grey90"))

table(rg_rp_se1$color_group)
# Plot the data
s22 <- rg_rp_se1 %>%
  ggplot(aes(y = rg, x = rp)) +
  
  # Plot grey points first
  geom_point(data = subset(rg_rp_se1, color_group == "grey90"), aes(colour = "grey90"), position = position_dodge(0.5), size = 3, show.legend = FALSE) +
  
  # Plot purple points on top (Both significant)
  geom_point(data = subset(rg_rp_se1, color_group == "b"), aes(colour = "b"), position = position_dodge(0.5), size = 3) +
  
  # Add labels only for traits_to_filter
  geom_label_repel(
    data = subset(rg_rp_se, (p1 %in% "ADIPO" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "ADIPO")& fdr_cat == "b"),
    aes(label = paste0(p1, "-", p2)),
    size = 4, box.padding = unit(0.5, "lines"), max.overlaps = 30,
    show.legend = FALSE
  ) +
  
  # geom_smooth(method = lm, formula = y ~ x, color = "grey5") +
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_vline(xintercept = 0, linetype = "dotted") + 
  scale_colour_manual(name = "Highlight",
                      labels = c("Both significant", "Other"), 
                      breaks = c("b", "grey90"), 
                      values = c(cbbPalette[1], "grey90")) +
  xlim(-0.7, 1) + ylim(-0.7, 1) + 
  xlab(bquote(r[p])) + ylab(bquote(r[g])) +
  theme(legend.position = "none", 
        legend.key = element_blank(), legend.text = element_text(size = 15), 
        legend.title = element_text(size = 15),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 20),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 20),
        panel.border = element_rect(colour = "black", fill = NA))

```

BIL
```{r}
cbbPalette <- c("#a161d1","#d16163","#618dd1","grey")
traits_to_filter <- c("BIL", "BMI", "FBS", "INSULIN", "WAIST", "LDL", "WBC")

# Create a new column to set the color for specific data points
rg_rp_se1 <- rg_rp_se %>%
  mutate(color_group = ifelse((p1 %in% "BIL" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "BIL") & fdr_cat == "b", fdr_cat, "grey90"))

table(rg_rp_se1$color_group)
# Plot the data
s23 <- rg_rp_se1 %>%
  ggplot(aes(y = rg, x = rp)) +
  
  # Plot grey points first
  geom_point(data = subset(rg_rp_se1, color_group == "grey90"), aes(colour = "grey90"), position = position_dodge(0.5), size = 3, show.legend = FALSE) +
  
  # Plot purple points on top (Both significant)
  geom_point(data = subset(rg_rp_se1, color_group == "b"), aes(colour = "b"), position = position_dodge(0.5), size = 3) +
  
  # Add labels only for traits_to_filter
  geom_label_repel(
    data = subset(rg_rp_se, (p1 %in% "BIL" & p2 %in% traits_to_filter|p1 %in% traits_to_filter & p2 %in% "BIL")& fdr_cat == "b"),
    aes(label = paste0(p1, "-", p2)),
    size = 4, box.padding = unit(0.5, "lines"), max.overlaps = 30,
    show.legend = FALSE
  ) +
  
  # geom_smooth(method = lm, formula = y ~ x, color = "grey5") +
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype = "dotted") + 
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_vline(xintercept = 0, linetype = "dotted") + 
  scale_colour_manual(name = "Highlight",
                      labels = c("Both significant", "Other"), 
                      breaks = c("b", "grey90"), 
                      values = c(cbbPalette[1], "grey90")) +
  xlim(-0.7, 1) + ylim(-0.7, 1) + 
  xlab(bquote(r[p])) + ylab(bquote(r[g])) +
  theme(legend.position = "none", 
        legend.key = element_blank(), legend.text = element_text(size = 15), 
        legend.title = element_text(size = 15),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 20),
        axis.text.x = element_text(size = 14),
        axis.title.y = element_text(size = 20),
        panel.border = element_rect(colour = "black", fill = NA))

```


## Slope
```{r}
summary(lm(rg~rp,data=rg_rp_se))$coefficients
#              Estimate Std. Error t value Pr(>|t|)    
# (Intercept) -0.009432   0.006144  -1.535    0.125    
# rp           0.634448   0.025926  24.472   <2e-16 ***
summary(lm(rp~rg,data=rg_rp_se))
```

# Save plots
```{r}
c <- corrplot(M, p.mat = M_pval, sig.level = 0.05,  insig = 'label_sig', pch.cex = 0.9, pch.col = 'grey20', type="full", method = 'square', diag = TRUE, order = 'hclust', addrect = 5, tl.col = 'black', tl.cex = 0.7)
```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_pleiotropy.pdf",height=4, width=8)
ggarrange(b,ncol = 1, #widths = c(2,1.5),
         labels = c("b"))
dev.off() 

```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/Fig2bc_10072024.pdf",height=7, width=20)
ggarrange(b,s2, ncol = 2, #widths = c(2,1.5),
         labels = c("b", "c"), font.label = list(size=20))
dev.off() 
```

```{r}
# pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_pleiotropy_merged.pdf",height=12, width=8)
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/SAIGE_kcps2_pleiotropy_merged_05102024.pdf",height=12, width=8)
ggarrange(a1,b, nrow = 2, #widths = c(2,1.5),
         labels = c("a", "b"))
dev.off() 

```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/kcps2_pleiotropy_merged_02112024.pdf",height=12, width=8)
ggarrange(a,b, nrow = 2, #widths = c(2,1.5),
         labels = c("a", "b"))
dev.off() 

```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/kcps2_pleiotropy_merged1_02112024.pdf",height=12, width=15)
ggarrange(a,b, nrow = 2, #widths = c(2,1.5),
         labels = c("a", "b"))
dev.off() 

```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/pleiotropy/kcps2_pleiotropy_merged_03012024.pdf",height=12, width=8)
ggarrange(a,b, nrow = 2, #widths = c(2,1.5),
         labels = c("a", "b"))
dev.off() 

```



#Old script

```{r}
#library("viridis")        
a <- ggplot(data = res1_long, aes(x = factor(description), y = N, fill = factor(report,levels = c("Known","Novel")))) + 
    geom_bar(stat="identity") + coord_flip() +  
   facet_grid(category~., scales = "free", space = "free",switch = "y", labeller = label_wrap_gen(width = 2, multi_line = TRUE)) +
    ylab("Number of significant loci") + xlab("") + # xlab("Trait")
  theme(strip.text.y = element_text(angle = 270, size = 4)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"), 
        legend.title =element_blank(), legend.text=element_text(size=15),
        axis.text.y = element_text(size = 10),
        axis.title=element_text(size=15))  +
  #scale_fill_discrete(name = "", labels = c("Novel ()","Previously Reported ()")) +
  scale_fill_brewer(palette = "Dark2") 

a+geom_text(aes(label=round(prop,2)), position=position_fill(.5))
a+geom_text(data=subset(res1_long, report=="Novel"), aes(label = round(perc,2), y = N - 3),
              position = position_dodge(width = 0.9)) 

  geom_text(aes(label = TotalDemand, y = TotalDemand - 3), nudge_x = c(0.22, -0.22)) + 

# theme_grey(base_size = 22)
#   #scale_color_viridis(discrete = TRUE, option = "D")+
#   # scale_fill_viridis(discrete = TRUE) 
# g+theme(axis.text=element_text(size=12),
#        axis.title=element_text(size=14,face="bold"))
```

```{r}
bim <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim")
snp <- bim$V2

# final 36 variables
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")

run_pleiotropy <- function(trait){
  gwas_fn <- paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/",trait,"/")
  file_list=list.files(path=gwas_fn, pattern = "*.linear")
  sumstats <- fread(file = paste0(gwas_fn,file_list[substr(file_list,1,5)=="Chr12"]))
  names(sumstats)[3] <- "SNP"
      
  sumstats2 <- sumstats %>%
    dplyr::filter(SNP %in% snp) %>%
    dplyr::rename(BP=POS, CHR=`#CHROM`) %>% #names(sumstats)[grepl("CHR",names(sumstats))]
    dplyr::mutate(#BP = as.numeric(BP),
                  logP=(-pnorm(abs(T_STAT), mean = 0, sd = 1, lower.tail = FALSE, log.p = T)-log(2))* log10(exp(1)),
                  sig=ifelse(logP>-log10(5e-8),1,0)) %>% #-log10(P) #x <- pnorm(-39, log.p = T); logP <- x * log10(exp(1))
    dplyr::select(BP, logP,sig)
  print(trait)
  #return(sumstats2$logP)
  return(sumstats2$sig)
}

trait=trait_list[1]
gwas_fn <- paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/",trait,"/")
  file_list=list.files(path=gwas_fn, pattern = "*.linear")
  sumstats <- fread(file = paste0(gwas_fn,file_list[substr(file_list,1,5)=="Chr12"]))
  names(sumstats)[3] <- "SNP"
      
  sumstats2 <- sumstats %>%
    dplyr::filter(SNP %in% snp) %>%
    dplyr::rename(BP=POS, CHR=`#CHROM`) %>% #names(sumstats)[grepl("CHR",names(sumstats))]
    dplyr::mutate(#BP = as.numeric(BP),
                  logP=(-pnorm(abs(T_STAT), mean = 0, sd = 1, lower.tail = FALSE, log.p = T)-log(2))* log10(exp(1)),
                  sig=ifelse(logP>-log10(5e-8),1,0)) %>% #-log10(P) #x <- pnorm(-39, log.p = T); logP <- x * log10(exp(1))
    dplyr::select(BP, logP,sig)

df_pleiotropy <- data.frame()
for(i in c(1:length(trait_list))){
  out_plt <- run_pleiotropy(trait_list[i])
  df_pleiotropy <- rbind(df_pleiotropy,out_plt)
}
df_pleiotropy <- as.data.frame(df_pleiotropy)
colnames(df_pleiotropy)<-sumstats2$BP
# for(i in c(1:ncol(df_pleiotropy))){
#   colnames(df_pleiotropy)[i] <- sumstats2$BP[i]+i
# }
rownames(df_pleiotropy)<-trait_list

```

```{r}
#install.packages("gplots")
# library("gplots")
# heatmap.2(as.matrix(df_pleiotropy), scale = "col", col = bluered(100), 
#           trace = "none", density.info = "none",dendrogram = "none")

```  
```{r}
library(reshape2); library(ggplot2);library(tidyr)
parRes_df <- melt(cbind(trait=rownames(df_pleiotropy), df_pleiotropy))
parRes_df2 <- parRes_df%>% 
      dplyr::mutate(category = case_when(
         (trait %in% c("BMI","HT","WAIST","WT")) ~ "Anthropometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardiovascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor marker",
              TRUE ~ "")) %>%
  dplyr::mutate(value1=ifelse(category=="Liver",value*2,
                       ifelse(category=="Thyroid hormone",value*3,
                       ifelse(category=="Tumor marker",value*4,
                       ifelse(category=="Kidney",value*5,
                       ifelse(category=="Hematological",value*6,data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABIAAAASCAYAAABWzo5XAAAAbElEQVR4Xs2RQQrAMAgEfZgf7W9LAguybljJpR3wEse5JOL3ZObDb4x1loDhHbBOFU6i2Ddnw2KNiXcdAXygJlwE8OFVBHDgKrLgSInN4WMe9iXiqIVsTMjH7z/GhNTEibOxQswcYIWYOR/zAjBJfiXh3jZ6AAAAAElFTkSuQmCC
                       ifelse(category=="Cardiovascular",value*7,
                       ifelse(category=="Anthropometry",value*8,
                       ifelse(category=="Lifestyle",value*9,value))))))))) %>%
  dplyr::mutate(category = factor(category, levels=unique(category)),
                trait = factor(trait, levels=c("BMI","HT","WAIST","WT","SBP","DBP","HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS","ALP","CREAT","URIC","COFFA","SMOKA_MOD","ALCO_AMOUNT","ALB","BIL","GGT","GOT","GPT","ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN","TSH","CEA"))) %>%
  arrange(category)

cbPalette <- c("white", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
plot_pleiotropy <- ggplot(parRes_df2, 
    aes(x = variable, y = trait, fill = factor(value1))) + 
    geom_tile() + 
    #scale_fill_manual(values=colors, aesthetics = c("colour", "fill")) + 
  scale_fill_manual(values=cbPalette) +
  theme(legend.position = "none") + xlab("") + ylab("")
```

```{r}
# final 36 variables
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")

df <- don %>%
  mutate(index=rownames(don), 
         chrpos = paste0(chr,"_",start,".",index), BPnum) %>%
  select(trait_list,chrpos) %>%
  remove_rownames() %>%
  as.tibble() %>%
  column_to_rownames("chrpos") %>%
  arrange("chrpos")

dft <- as.data.frame(t(df))
df1 <- melt(cbind(trait=rownames(dft), dft))
df2 <- df1 %>% 
      dplyr::mutate(category = case_when(
         (trait %in% c("BMI","HT","WAIST","WT")) ~ "Anthropometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardiovascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor marker",
              TRUE ~ "")) %>%
  dplyr::mutate(value1=ifelse(category=="Liver",value*2,
                       ifelse(category=="Thyroid hormone",value*3,
                       ifelse(category=="Tumor marker",value*4,
                       ifelse(category=="Kidney",value*5,
                       ifelse(category=="Hematological",value*6,
                       ifelse(category=="Cardiovascular",value*7,
                       ifelse(category=="Anthropometry",value*8,
                       ifelse(category=="Lifestyle",value*9,value))))))))) %>%
  dplyr::mutate(category = factor(category, levels=unique(category)),
                trait = factor(trait, levels=c("BMI","HT","WAIST","WT","SBP","DBP","HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS","ALP","CREAT","URIC","COFFA","SMOKA_MOD","ALCO_AMOUNT","ALB","BIL","GGT","GOT","GPT","ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN","TSH","CEA")))

cbPalette <- c("white","#CC6677" ,"#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#AA4499","#888888")

ggplot(df2, 
    aes(x = variable, y = trait, fill = factor(value1))) + 
    geom_tile() + 
    #scale_fill_manual(values=colors, aesthetics = c("colour", "fill")) + 
    scale_fill_manual(values=cbPalette) +
  theme(legend.position = "none") + xlab("") + ylab("")
```

```{r}


# a2 <- ggplot(data = res1_long) +
#   aes(x = factor(description), fill = factor(report,levels = c("Known","Novel")), weight = N, by = factor(description)) +
  #geom_bar(position = "fill") +

# a2 <- ggplot(data = res1_long, aes(x = factor(description), weight = N), by = factor(description)) +
#   geom_bar(fill = "grey", alpha = .5) +
#   geom_bar(data = subset(res1_long, report=="Novel"), fill = sort(data$col_cat), alpha = .8) +
#   
#   geom_text(stat = "prop", position = position_fill(.5)) + 
#   coord_flip() +  
#    #facet_grid(category~., scales = "free", space = "free",switch = "y", labeller = label_wrap_gen(width = 2, multi_line = TRUE)) +
#     ylab("Fraction of novel loci") + xlab("") + 
#   theme(strip.text.y = element_text(angle = 270, size = 4)) +
#   theme(panel.grid.major = element_blank(), 
#         panel.grid.minor = element_blank(), 
#         panel.background = element_blank(), 
#         axis.line = element_line(colour = "black"), 
#         legend.title =element_blank(), legend.text=element_text(size=15),
#         axis.text.y = element_text(size = 10, colour = sort(res1_long$col_cat)),
#         axis.title=element_text(size=15),
#         legend.position = "none")  
  #scale_fill_brewer(palette = "Dark2") 
  #scale_fill_manual(labels = c("Known","Novel"), values = c("steelblue", "gold"))

# colfill <- c("#DB9D85","#DB9D85","#DB9D85","#DB9D85","#BFAA67","#BFAA67","#BFAA67","#BFAA67","#BFAA76","#BFAA76","#BFAA67","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B66E","#93B666","#93B66E","#5CBD92","#5CBD92","#38BDBB","#38BDBB","#38BDBB","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#6CB4D9","#ACA4E2","#D596D2","#E494AF","#E494AF","#E494AF")
```
