---
title: "locuszoom_susie_coloc"
author: "Yon Ho Jee"
date: "2024-01-28"
output: html_document
---
```{r}
library(data.table)
library(dplyr)
library(ggplot2)
library(cowplot)
library(tidyverse)
```
```{r warning=FALSE}
library(data.table)
library(tidyverse)
library(gridExtra)
library(grid)
#library(TigR)
library(ggrepel)
library(ggpubr)
```
```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
library(data.table);library(dplyr)
library(susieR)
#library(ieugwasr)
library(Rfast)
set.seed(1)
```

# =======
# SuSiE
# =======
## Get GWAS
```{r}
trait="ALCO_AMOUNT"
get_gwas_rs671=function(trait){
  
  bim <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim"),data.table = F)
  snp <- bim$V2

  gwas_fn = paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/SAIGE/loco/")
  gwas0 <- fread(paste0(gwas_fn,"SAIGE_",trait,"_INFO.txt.gz"), data.table = F) 
  gwas1 <- gwas0 %>%
    #dplyr::select(chr, pos, snp, p, beta, se) %>%
    #dplyr::filter(CHR==12 & POS<=(112241766+500000) & POS>=(112241766-500000)) #rs671+/-500kb (500kb=500000bp)
    dplyr::filter(MarkerID %in% snp)
  print(trait)
  return(gwas1)
}
```

## Revision: just get effects of rs671 on coloc.traits
```{r}
trait="ALCO_AMOUNT"
trait_list_coloc <- c("ALCO_AMOUNT","GOT","GPT","GGT","SBP","DBP","COFFA","TG")
gwas0_df <- data.frame()
for(trait in trait_list_coloc){
  
  snp <- "rs671"

  gwas_fn = paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/SAIGE/loco/")
  gwas0 <- fread(paste0(gwas_fn,"SAIGE_",trait,"_INFO.txt.gz"), data.table = F) %>%
    dplyr::filter(MarkerID %in% snp) %>%
    mutate(Trait=trait)
  print(gwas0)
  gwas0_df <- rbind(gwas0_df,gwas0)
}
```

```{r}
run_susie <- function(trait){

  # Merge gwas and extracting SNPs
  sumstats <- get_gwas_rs671(trait)
      
  # Z-score
  z_scores <- sumstats$BETA / sumstats$SE
  
  # LD matrix
  R_ref <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr12_kcps2_r.ld.gz"))
  R_ref <- as.matrix(R_ref)
  rownames(R_ref)<-colnames(R_ref)
  
  ## Run SuSiE
  fitted_rss3 <- susieR::susie_rss(z_scores, R_ref, n = sumstats$N[1], L = 10, refine = TRUE)
  print(paste0("SuSiE converged: ",fitted_rss3$converged))
  
  susie_plot(fitted_rss3, y="PIP")
  #ggsave(paste0("/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/plot/susie_plot_",trait,".png"), width = 15, height = 7)

  print(trait)
  return(fitted_rss3)  
}
```

## Run susie
```{r}
# final 36 variables
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HEIGHT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")
```

```{r}
all_trait_output <- list()               # Create empty list
for(k in 1:length(trait_list)){
  all_trait_output[[trait_list[k]]] <- run_susie(trait_list[k])
}
saveRDS(all_trait_output,"/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_susie_output")
```

## Collect outputs
```{r}
#all_trait_output <- readRDS("/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_susie_output")

pip_res <- c()
for(k in 1:length(trait_list)){
  bim <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim"),data.table = F)
  snp <- bim$V2

  mean_pip0 <- mean(all_trait_output[[trait_list[k]]]$pip)
  #count_pip1 <- sum(all_trait_output[[trait_list[k]]]$pip>0)
  cs <- all_trait_output[[trait_list[k]]]$sets$cs
  N_cs <- length(cs)
  Total_size_cs <- sum(length(cs$L1),length(cs$L2),length(cs$L3),length(cs$L4),length(cs$L5),length(cs$L6),length(cs$L7),length(cs$L8),length(cs$L9),length(cs$L10), na.rm = TRUE)
  N_cs_1 <- sum(length(cs$L1)==1,length(cs$L2)==1,length(cs$L3)==1,length(cs$L4)==1,length(cs$L5)==1,length(cs$L6)==1,length(cs$L7)==1,length(cs$L8)==1,length(cs$L9)==1,length(cs$L10)==1, na.rm = TRUE)
  size_cs <- paste(c(length(cs$L1),length(cs$L2),length(cs$L3),length(cs$L4),length(cs$L5),length(cs$L6),length(cs$L7),length(cs$L8),length(cs$L9),length(cs$L10)),sep=",",collapse = ",")

  SNP <- paste(snp[c(cs$L1,cs$L2,cs$L3,cs$L4,cs$L5,cs$L6,cs$L7,cs$L8,cs$L9,cs$L10)],sep=",",collapse = ",")
  Is_rs671 <- str_detect(SNP, "rs671")
  snp_index <- which(snp=="rs671")
  PIP_rs671 <- all_trait_output[[trait_list[k]]]$pip[snp_index] #rs671: 622
 
  #mean_pip1 <- cbind(trait_list[k], mean_pip0, count_pip1, nrow(snp), N_cs, size_cs, SNP)
  mean_pip1 <- cbind(trait_list[k], mean_pip0, length(snp), N_cs, N_cs_1, Total_size_cs, size_cs, SNP, Is_rs671, PIP_rs671)
  
  pip_res <- rbind(pip_res, mean_pip1)
}
pip_res <- as.data.frame(pip_res)
colnames(pip_res) <- c("trait","mean_PIP", "Number of SNPs fine-mapped","Number of CS", "Number of CS with 1 variant","Total size of CS (#variants)","Size of each CS","SNPs in cs","Is_rs671","PIP_rs671")

```
```{r}
write.csv(pip_res,"/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_mean_pip_rs671.csv",row.names = FALSE)
```

```{r}
pip_res <- read.csv("/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_mean_pip_rs671.csv")
```

```{r}
summary(as.numeric(pip_res$Number.of.CS))
  #  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  # 0.000   1.000   1.000   1.722   2.000  10.000
sum(as.numeric(pip_res$Number.of.CS))
#62
summary(as.numeric(pip_res$`Mean CS`))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1.00    5.00    8.00   42.04   21.50  225.00 
#if divided by L=10
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.100   0.500   0.900   8.428   5.800  57.000 

summary(as.numeric(pip_res$Total.size.of.CS...variants.))
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 0.00    1.00    6.50   86.47  102.50  470.00 
   
sum(pip_res$Is_rs671=="TRUE")
#23

pip_res$PIP_rs671[order(-as.numeric(pip_res$PIP_rs671))][1:10]
sum(as.numeric(pip_res$PIP_rs671)>0.9)
#8
trait_list_fm <- pip_res$trait[as.numeric(pip_res$PIP_rs671)>0.9]
# [1] "TG"          "GOT"         "GPT"         "GGT"         "SBP"         "DBP"        
# [7] "ALCO_AMOUNT" "COFFA" 

View(pip_res[as.numeric(pip_res$PIP_rs671)>0.9,])

sum(as.numeric(pip_res$Total.size.of.CS...variants.))/sum(as.numeric(pip_res$Number.of.CS)) #50.20968
min(as.numeric(pip_res$Total.size.of.CS...variants.)) #0
max(as.numeric(pip_res$Total.size.of.CS...variants.)) #470

pip_res_rs671 <- pip_res[as.numeric(pip_res$PIP_rs671)>0.9,]
sum(as.numeric(pip_res_rs671$Total.size.of.CS...variants.))/sum(as.numeric(pip_res_rs671$Number.of.CS)) #20.13636
min(as.numeric(pip_res_rs671$Total.size.of.CS...variants.)) #1
max(as.numeric(pip_res_rs671$Total.size.of.CS...variants.)) #262

sum(as.numeric(pip_res$Number.of.CS.with.1.variant)) #21
```

#=============
# coloc
#=============
```{r}  
library(coloc)
bim <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim")
snp <- bim$V2

R_ref <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr12_kcps2_r.ld.gz"))
R_ref <- as.matrix(R_ref)
rownames(R_ref) <- snp
colnames(R_ref) <- rownames(R_ref)
```

```{r}
trait_list_coloc <- c("ALCO_AMOUNT","GOT","GPT","GGT","SBP","DBP","COFFA","TG")

coloc_susie_output <- list()               # Create empty list
all_pp4_rs671 <- c()
run_coloc_susie <- function(trait2){
  trait1="ALCO_AMOUNT"
  #trait2="GGT"
  
  gwas1 <- get_gwas_rs671(trait=trait1) %>%
    dplyr::rename(rsid=MarkerID, maf=AF_Allele2)
  colnames(gwas1) <- tolower(names(gwas1))
  
  gwas2 <- get_gwas_rs671(trait=trait2) %>%
    dplyr::rename(rsid=MarkerID, maf=AF_Allele2)
  colnames(gwas2) <- tolower(names(gwas2))
  
  
  #merge
  input <- merge(gwas1, gwas2, by="rsid", all=FALSE, suffixes=c("_gwas1","_gwas2")) #5596831
  #input <- input[input$pvalue_gwas1<5e-8 & input$pvalue_gwas2<5e-8,] #442
  
  #reform
  coloc_list <- list()
  temp <- subset(input, select=c("beta_gwas1","se_gwas1","rsid","pos_gwas1","maf_gwas1"))
  temp <- temp[!duplicated(temp),]
  temp$se_gwas1 <- temp$se_gwas1**2
  colnames(temp) <- c("beta","varbeta","snp","position","MAF")
  coloc_list[["gwas1"]] <- as.list(temp)
  coloc_list[["gwas1"]]["type"] <- "quant"
  coloc_list[["gwas1"]]["N"] <- gwas1$n[1]
  coloc_list[["gwas1"]]["LD"] <- list(R_ref)
  
  #temp <- merge_dat[merge_dat$tissue==loop_tissue,]
  temp <- subset(input, select=c("beta_gwas2","se_gwas2","rsid","pos_gwas2","maf_gwas2"))
  temp <- temp[!duplicated(temp$rsid),]
  temp$se_gwas2 <- temp$se_gwas2**2
  temp <- temp[!is.na(temp$se_gwas2),]
  colnames(temp) <- c("beta","varbeta","snp","position","MAF")
  coloc_list[["gwas2"]] <- as.list(temp)
  coloc_list[["gwas2"]]["type"] <- "quant"
  coloc_list[["gwas2"]]["N"] <- gwas2$n[1]
  coloc_list[["gwas2"]]["LD"] <- list(R_ref)
  
  D3=coloc_list$gwas1
  D4=coloc_list$gwas2

  my.res <- coloc.abf(dataset1=D3,dataset2=D4)
  
  S3=coloc::runsusie(D3,n=gwas1$n[1])
  S4=coloc::runsusie(D4,n=gwas2$n[1])
  
  if(requireNamespace("susieR",quietly=TRUE)) {
    susie.res=coloc.susie(S3,S4)
    print(susie.res$summary)
    print(susie.res$results)
  }
  
  pp4_rs671 <- susie.res$results$SNP.PP.H4.row1[which(susie.res$results$snp=="rs671")]
  
  return(list(my.res,pp4_rs671))

}

for(k in 2:length(trait_list_coloc)){
  coloc_susie_output[[trait_list_coloc[k]]] <- run_coloc_susie(trait_list_coloc[k])[[1]]
  all_pp4_rs671 <- c(all_pp4_rs671, run_coloc_susie(trait_list_coloc[k])[[2]])
}
```
```{r}
all_pp4_rs671
#[1] 1 1 1 1 1 1
```

#=============
# Locuszoom
#=============
https://yosuketanigawa.com/posts/2020/07/small-values-in-R/
```{r warning=FALSE}
library(data.table)
library(dplyr)
library(ggplot2)
library(cowplot)
library(tidyverse)
library(gridExtra)
```
```{r warning=FALSE}
library(locuszoomr)
library(grid)
```

Run All-in-one using function
```{r}
gene_of_interest = "ALDH2"
get_input_locuszoom <- function(sumstats_path_filename=NULL,
                                trait=NULL, chr=NULL,
                         bim_path_filename, 
                         ldm_path_filename, 
                         snp_of_interest=NULL, #default is most significant SNP
                         gene_of_interest){
  # 1. Load sumstat & bim file used to create LD matrix
  #sumstats_path_filename <- "~/GLP1R.csv"
  if(!is.null(sumstats_path_filename)){
    sumstats <- fread(sumstats_path_filename, data.table = F)
    sumstats$Z <- sumstats$BETA/sumstats$SE
    } else {
      gwas_fn = paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/SAIGE/loco/")
      sumstats <- fread(paste0(gwas_fn,"SAIGE_",trait,"_INFO.txt.gz"), data.table = F) 
      }

  #bim_path_filename <- "/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_GLP1R.bim"
  bim <- fread(bim_path_filename)
  snp <- bim$V2
  
  #3. Match SNP list between sumstats and LD matrix
  sumstats2 <- sumstats %>%
    dplyr::filter(MarkerID %in% snp) %>%
    # separate(
    #     p.value, c("P_base", "P_exp"),
    #     sep = "E", remove = F, fill = "right"
    # ) %>%
    # replace_na(list(P_exp = "0")) %>%
    # mutate(log10P = log10(as.numeric(P_base)) + as.numeric(P_exp)) %>%
    dplyr::rename(SNP=MarkerID) %>% #names(sumstats)[grepl("CHR",names(sumstats))]
    dplyr::select(CHR, POS, SNP, neglogP)
  
  # 2. Load LD matrix
  #ldm_path_filename <- "/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_r2_GLP1R.ld.gz"
  R_ref <- fread(ldm_path_filename)
  R_ref <- as.matrix(R_ref)
  colnames(R_ref) <- snp
  rownames(R_ref) <- snp
  
  # convert LD matrix to long format
  R_ref_ld <- as.data.frame(as.table(cor(R_ref))) 
  R_ref_ld <- as.data.frame(cbind(R_ref_ld,rep(sumstats2$CHR[1],nrow(R_ref)*nrow(R_ref)),rep(sumstats2$CHR[1],nrow(R_ref)*nrow(R_ref)),rep(sumstats2$POS,each=nrow(R_ref)),rep(sumstats2$POS,each=nrow(R_ref))))
  R_ref_ld$R2<-R_ref_ld$Freq^2
  R_ref_ld <- R_ref_ld[,-3]
  names(R_ref_ld)<-c("SNP_A","SNP_B","CHR_A","CHR_B","POS_A","POS_B","R2")
  # CHR_A         POS_A         SNP_A  CHR_B         POS_B         SNP_B           R2 
  
  # extract the most significant SNP from LD long format
  if(is.null(snp_of_interest)){
    most_sig_snp <- sumstats2$SNP[order(-sumstats2$neglogP)][1]
  } else most_sig_snp <- snp_of_interest
  
  gene <- gene_of_interest
  
  R_ref_ld <- R_ref_ld[R_ref_ld$SNP_A==most_sig_snp,]
  
  R_ref_ld2 <- R_ref_ld %>%
    dplyr::rename(rsid=SNP_B, r2=R2)
  
  sumstats3 <- sumstats %>%
    dplyr::filter(MarkerID %in% snp) %>%
    dplyr::rename(rsid=MarkerID) %>%
    dplyr::select(CHR, POS, rsid, neglogP)
  
  gwas_r2 <- left_join(sumstats3, R_ref_ld2, by="rsid") %>%
    dplyr::select(CHR, POS, rsid, neglogP, r2) 

  return(list(sumstats2, R_ref_ld, most_sig_snp, gene,
              gwas_r2))
}

```

```{r}
get_loc_locuszoomr <- function(get_input_list, window){
  
  if(!is.null(get_input_list$sumstats_path_filename)){
    sumstats_path_filename = get_input_list$sumstats_path_filename
    trait = NULL; chr = NULL
    } else {
      sumstats_path_filename = NULL
      trait = get_input_list$trait; chr = get_input_list$chr
    }
  bim_path_filename = get_input_list$bim_path_filename
  ldm_path_filename = get_input_list$ldm_path_filename
  snp_of_interest = get_input_list$snp_of_interest
  gene_of_interest = get_input_list$gene_of_interest

  res <- get_input_locuszoom(sumstats_path_filename = sumstats_path_filename,
                             trait = trait, chr = chr,
                             bim_path_filename = bim_path_filename,
                             ldm_path_filename = ldm_path_filename,
                             snp_of_interest = snp_of_interest,
                             gene_of_interest = gene_of_interest)
  
  gwas_r2 <- res[[5]]
  most_sig_snp <- res[[3]]
  gene <- res[[4]]
  
  
  if (require(EnsDb.Hsapiens.v75)) {
    loc <- locus(gwas_r2, gene = gene, flank = window, #1e5
             LD = "r2", ens_db = "EnsDb.Hsapiens.v75", index_snp = most_sig_snp, yvar = "neglogP")
    #locus_ggplot(loc, labels = "index", label_x = c(4, -5), ylab = bquote("-log"[10]*"P-value"))
    locus_plot(loc, labels = "index", label_x = c(4, -5), ylab="-log10(P-value)",
  cex = 1,
  cex.axis = 0.9,
  cex.lab = 1.5,
  cex.text = 0.7)
    #title(ylab = bquote("-log"[10]*"P-value"))
  }
  
  return(loc)
}
```

```{r}
get_loc_locuszoomr_coloc <- function(get_input_list, window){
  
  if(!is.null(get_input_list$sumstats_path_filename)){
    sumstats_path_filename = get_input_list$sumstats_path_filename
    trait = NULL; chr = NULL
    } else {
      sumstats_path_filename = NULL
      trait = get_input_list$trait; chr = get_input_list$chr
    }
  bim_path_filename = get_input_list$bim_path_filename
  ldm_path_filename = get_input_list$ldm_path_filename
  snp_of_interest = get_input_list$snp_of_interest
  gene_of_interest = get_input_list$gene_of_interest

  res <- get_input_locuszoom(sumstats_path_filename = sumstats_path_filename,
                             trait = trait, chr = chr,
                             bim_path_filename = bim_path_filename,
                             ldm_path_filename = ldm_path_filename,
                             snp_of_interest = snp_of_interest,
                             gene_of_interest = gene_of_interest)
  
  gwas_r2 <- res[[5]]
  most_sig_snp <- res[[3]]
  gene <- res[[4]]
  
  
  if (require(EnsDb.Hsapiens.v75)) {
    loc <- locus(gwas_r2, gene = gene, flank = window, #1e5
             LD = "r2", ens_db = "EnsDb.Hsapiens.v75", index_snp = most_sig_snp, yvar = "neglogP")
    #locus_ggplot(loc, labels = "index", label_x = c(4, -5), ylab = bquote("-log"[10]*"P-value"))
    locus_plot(loc, labels = "index", label_x = c(4, -5), ylab="-log10(P)")
    #title(ylab = bquote("-log"[10]*"P-value"))
  }
  
  return(loc)
}
```

tmp
```{r}
p <- gg_scatter(loc,index_snp = loc$index_snp)
title(ylab = bquote("-log"[10]*"P-value"))

# Create the plot
plot <- locus_ggplot(loc, ylab = bquote("-log"[10]*"P-value"))

# Extract data for the index SNP to find its position
index_snp_data <- subset(loc$data, rsid == loc$index_snp)

# Add a label for the index SNP using geom_text or geom_label
plot <- plot +
    geom_text(data = index_snp_data, aes(x = POS, y = neglogP, label = rsid))  # Customize position and style as needed

# Print the modified plot with SNP label
print(plot)

plot <- locus_ggplot(loc, ylab = bquote("-log"[10]*"P-value")) +
    geom_label_repel(data = subset(loc$data, rsid == loc$index_snp), aes(x = POS, y = neglogP, label = rsid),
    nudge_y = 1, # Optional: nudge the label vertically for better spacing
    nudge_x = 0) # Optional: nudge the label horizontally

locus_ggplot(loc, ylab = bquote("-log"[10]*"P-value")) +
    annotate(
    "text", label = "plot mpg vs. wt",
    x = 2, y = 15, size = 8, colour = "red"
  )
```


Format input_list 
```{r}
GWAS_input_list_ft <- function(gene,trait,chr,eQTL_snp){
  input_list2 <- list(trait=trait,chr=chr,
                      bim_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr",chr,".bim"), 
                     ldm_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr",chr,"_kcps2_r.ld.gz"),
                       snp_of_interest=eQTL_snp, 
                    gene_of_interest=gene)
  return(input_list2)
}
```

```{r}
gene="ALDH2"
trait1="ALCO_AMOUNT"
trait2="GOT"
trait3="GPT"
trait4="GGT"
trait5="SBP"
trait6="DBP"
trait7="COFFA"
trait8="TG"

chr=12
eQTL_snp="rs671"
eQTL_snp=NULL
kcps2_gwas_input_list1 <- GWAS_input_list_ft(gene,trait1,chr,eQTL_snp)
kcps2_gwas_input_list2 <- GWAS_input_list_ft(gene,trait2,chr,eQTL_snp)
kcps2_gwas_input_list3 <- GWAS_input_list_ft(gene,trait3,chr,eQTL_snp)
kcps2_gwas_input_list4 <- GWAS_input_list_ft(gene,trait4,chr,eQTL_snp)
kcps2_gwas_input_list5 <- GWAS_input_list_ft(gene,trait5,chr,eQTL_snp)
kcps2_gwas_input_list6 <- GWAS_input_list_ft(gene,trait6,chr,eQTL_snp)
kcps2_gwas_input_list7 <- GWAS_input_list_ft(gene,trait7,chr,eQTL_snp)
kcps2_gwas_input_list8 <- GWAS_input_list_ft(gene,trait8,chr,eQTL_snp)
```

```{r}
loc <- get_loc_locuszoomr(get_input_list = kcps2_gwas_input_list1, window=5e5)
#loc <- link_recomb(loc, genome = "hg19")
loc2 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list2, window=5e5)
#loc2 <- link_recomb(loc2, genome = "hg19")
loc3 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list3, window=5e5)
#loc3 <- link_recomb(loc3, genome = "hg19")
loc4 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list4, window=5e5)
#loc4 <- link_recomb(loc4, genome = "hg19")
loc5 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list5, window=5e5)
#loc5 <- link_recomb(loc5, genome = "hg19")
loc6 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list6, window=5e5)
#loc6 <- link_recomb(loc6, genome = "hg19")
loc7 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list7, window=5e5)
#loc7 <- link_recomb(loc7, genome = "hg19")
loc8 <- get_loc_locuszoomr_coloc(get_input_list = kcps2_gwas_input_list8, window=5e5)
#loc8 <- link_recomb(loc8, genome = "hg19")
```

```{r}
# pdf("~/locuszoom_rs671_v2.pdf", width = 9, height = 12)
# oldpar <- set_layers(7)
# scatter_plot(loc, xticks = FALSE, labels = "index",legend_pos = "topright")
# legend("topleft", legend=c(trait1), bty = "n")
# scatter_plot(loc2, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait2), bty = "n")
# scatter_plot(loc3, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait3), bty = "n")
# scatter_plot(loc4, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait4), bty = "n")
# scatter_plot(loc5, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait5), bty = "n")
# scatter_plot(loc6, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait6), bty = "n")
# scatter_plot(loc7, xticks = FALSE, legend = FALSE)
# legend("topleft", legend=c(trait7), bty = "n")
# genetracks(loc)
# par(oldpar)  # revert par() settings
# dev.off()
```

```{r}
susie_output <- all_trait_output[["ALCO_AMOUNT"]]
# snp[susie_output$sets$cs$L1] #rs671
# snp[susie_output$sets$cs$L3] #rs141043717
# snp[susie_output$sets$cs$L4] #rs147992802
# snp[susie_output$sets$cs$L7] #rs61055528
# snp[susie_output$sets$cs$L10] #rs149178839

snp[susie_output$sets$cs$L1] #rs671
susie_output$pip[susie_output$sets$cs$L1]

snp[susie_output$sets$cs$L2] #rs555501971
susie_output$pip[susie_output$sets$cs$L2]

snp[susie_output$sets$cs$L4] #rs141043717
susie_output$pip[susie_output$sets$cs$L4]

snp[susie_output$sets$cs$L5] #rs61055528
susie_output$pip[susie_output$sets$cs$L5]

snp[susie_output$sets$cs$L6] #rs149178839
susie_output$pip[susie_output$sets$cs$L6]

snp[susie_output$sets$cs$L7] #rs11066008
susie_output$pip[susie_output$sets$cs$L7]

snp[susie_output$sets$cs$L8] #rs550463060
susie_output$pip[susie_output$sets$cs$L8]

```

#===================
# Plot CS manually
#===================
```{r}
all_trait_output <- readRDS("/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_susie_output")

# trait_list_coloc=c("ALCO_AMOUNT","GGT","SBP","DBP","COFFA","FBS","TG")
trait_list_coloc <- c("ALCO_AMOUNT","GOT","GPT","GGT","SBP","DBP","COFFA","TG")

# all_trait_output <- list()
# for(k in 1:length(trait_list_coloc)){
#   all_trait_output[[trait_list_coloc[k]]] <- run_susie(trait_list_coloc[k])
# }

bim <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim")
snp <- bim$V2
pos <- bim$V4  

# snp[all_trait_output$ALCO_AMOUNT$sets$cs$L1,] #rs671
# all_trait_output$ALCO_AMOUNT$pip[all_trait_output$ALCO_AMOUNT$sets$cs$L1]
# all_trait_output$ALCO_AMOUNT$pip[all_trait_output$ALCO_AMOUNT$sets$cs$L9]

make_locuszoom_susie_coloc_alco <- function(susie_output){
  cs_plt <- data.frame()
  for(i in c(1:length(susie_output$sets$cs))){
    cs <- rep(susie_output$sets$cs_index[[i]], length(susie_output$sets$cs[[i]]))
    pip <- susie_output$pip[susie_output$sets$cs[[i]]]
    s <- snp[susie_output$sets$cs[[i]]]
    p <- pos[susie_output$sets$cs[[i]]]
    out <- cbind(s,p,pip,cs)
    cs_plt <- rbind(cs_plt, out)
  }
  names(cs_plt) <- c("SNP","POS","PIP","CS")
  cs_plt <- cs_plt %>%
    mutate(CS = factor(CS,levels=c("1","2","3","4","5","6","7","8","9","10")),
           POS = as.numeric(POS), PIP = as.numeric(PIP))
  
  cs_plt1 <- cs_plt %>%
    ggplot(aes(POS, PIP, col=factor(CS)))+ 
    #geom_point(position = position_dodge(0.5), size = 5)+
    geom_point() +
    # geom_label_repel(
    #   aes(label=trait), size=4, box.padding = unit(0.5, "lines")
    # )+
    ggpubr::theme_pubr(border = TRUE) +
    geom_abline(intercept = 0, slope = 1, linetype="dotted") + 
    xlim(112241766-530000, 112241766+500000) + xlab("") + ylab("PIP") + 
    #112241766+500000, 112241766-500000
    scale_y_continuous(limits = c(0,1), breaks=c(0,0.5,1)) +
    theme(legend.position = c(0.9,0.5), 
          axis.text.x=element_blank(), axis.ticks.x=element_blank(),
          legend.title = element_text(size = 8),
          legend.text = element_text(size = 8),
          legend.key = element_rect(colour = NA, fill = NA), 
          legend.background = element_rect(color = NA, fill='white'), #transparent legend bg
          legend.box.background = element_rect(fill='white'), #transparent legend panel
          legend.spacing.x = unit(0, 'cm'), legend.spacing.y = unit(0, 'cm'),
          axis.title.y = element_text(size = 20)) + 
    guides(col = guide_legend(ncol = 3, title="95% CS"))
  
  return(cs_plt1)
}

susie_output <- all_trait_output[["ALCO_AMOUNT"]]
cs_plt1 <- make_locuszoom_susie_coloc_alco(susie_output)
  
make_locuszoom_susie_coloc <- function(susie_output){
  cs_plt <- data.frame()
  for(i in c(1:length(susie_output$sets$cs))){
    cs <- rep(susie_output$sets$cs_index[[i]], length(susie_output$sets$cs[[i]]))
    pip <- susie_output$pip[susie_output$sets$cs[[i]]]
    s <- snp[susie_output$sets$cs[[i]]]
    p <- pos[susie_output$sets$cs[[i]]]
    out <- cbind(s,p,pip,cs)
    cs_plt <- rbind(cs_plt, out)
  }
  names(cs_plt) <- c("SNP","POS","PIP","CS")
  cs_plt <- cs_plt %>%
    mutate(CS = factor(CS,levels=c("1","2","3","4","5","6","7","8","9","10")),
           POS = as.numeric(POS), PIP = as.numeric(PIP))
  
  cs_plt1 <- cs_plt %>%
    ggplot(aes(POS, PIP, col=factor(CS)))+ 
    #geom_point(position = position_dodge(0.5), size = 5)+
    geom_point() +
    # geom_label_repel(
    #   aes(label=trait), size=4, box.padding = unit(0.5, "lines")
    # )+
    ggpubr::theme_pubr(border = TRUE) +
    geom_abline(intercept = 0, slope = 1, linetype="dotted") + 
    xlim(112241766-530000, 112241766+500000) + xlab("") + ylab("PIP") + 
    #112241766+500000, 112241766-500000
    scale_y_continuous(limits = c(0,1), breaks=c(0,0.5,1)) +
    theme(legend.position = c(0.85,0.5), 
          axis.text.x=element_blank(), axis.ticks.x=element_blank(),
          legend.title = element_text(size = 8),
          legend.text = element_text(size = 8),
          legend.key = element_rect(colour = NA, fill = NA), 
          legend.background = element_rect(color = NA, fill='transparent'), #transparent legend bg
          legend.box.background = element_rect(color = NA, fill='transparent'), #transparent legend panel
          legend.spacing.x = unit(0, 'cm'), legend.spacing.y = unit(0, 'cm')
          ) + 
    guides(col = guide_legend(ncol = 3, title="95% CS"))
  
  return(cs_plt1)
}

for(j in 2:length(trait_list_coloc)){
  susie_output <- all_trait_output[[trait_list_coloc[j]]]
  assign(paste0("cs_plt",j),
         make_locuszoom_susie_coloc(susie_output))
  print(paste0("cs_plt",j));print(trait_list_coloc[j])
}

```

```{r}
#ggarrange(locus_ggplot(loc),cs_plt1,nrow=2)
lx <- loc$data$pos[loc$data$rsid == loc$index_snp]
ly <- loc$data$logP[loc$data$rsid == loc$index_snp]

grob1 <- grobTree(textGrob("Alcohol amount", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob(loc$index_snp, x=0.5, y=0.98, hjust=0,
  gp=gpar(col="black", fontsize=10)))

p1 <- gg_scatter(loc, xticks = FALSE, legend_pos = "topright", border = TRUE) + annotation_custom(grob1)
g <- gg_genetracks(loc,maxrows = 3)

pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/locuszoom/susie_coloc/SAIGE_alco_susie_coloc.pdf")
plot_grid(p1, cs_plt1, g, ncol = 1, rel_heights = c(2, 2, 1), align = "v")
dev.off()
```


```{r}
grob1 <- grobTree(textGrob("Alcohol amount", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob(loc$index_snp, x=0.5, y=0.98, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob2 <- grobTree(textGrob("GOT", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob3 <- grobTree(textGrob("GPT", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob4 <- grobTree(textGrob("GGT", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob5 <- grobTree(textGrob("SBP", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob6 <- grobTree(textGrob("DBP", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob7 <- grobTree(textGrob("Coffee intake", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
grob8 <- grobTree(textGrob("TG", x=0.01,  y=0.80, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")))
```

```{r}
p1 <- gg_scatter(loc, xticks = FALSE, legend_pos = "topright", border = TRUE, ylab="-log10(P-value)", cex.lab = 2) + annotation_custom(grob1) 
p2 <- gg_scatter(loc2, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob2)
p3 <- gg_scatter(loc3, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob3)
p4 <- gg_scatter(loc4, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob4)
p5 <- gg_scatter(loc5, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob5)
p6 <- gg_scatter(loc6, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob6)
p7 <- gg_scatter(loc7, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob7)
p8 <- gg_scatter(loc8, xticks = FALSE, legend_pos = NULL, border = TRUE, ylab="-log10(P)") + annotation_custom(grob8)

g <- gg_genetracks(loc,maxrows = 3)
```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/locuszoom/susie_coloc/SAIGE_all_susie_coloc.pdf", width = 9, height = 12)
#plot_grid(p1, cs_plt1, p2, p3, p4, p5, p6, p7, g, ncol = 1, rel_heights = c(2, 2, 1), align = "v")
# plot_grid(p1, cs_plt1, p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, g, ncol = 1, rel_heights = c(2, 2, 1), align = "v")
# plot_grid(p2, p3, p4, p5, p6, p7, g, ncol = 1, rel_heights = c(2, 2, 2, 2, 2, 2, 1), align = "v")
# plot_grid(p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, g, ncol = 1, rel_heights = c(2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 1), align = "v")

#plot_grid(p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, g, ncol = 1, rel_heights = c(2, 1.5, 2, 1.5, 2, 1.5, 2, 1.5, 2, 1.5, 2), align = "v")
plot_grid(p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, p8, cs_plt8, g, ncol = 1, align = "v")
dev.off()
```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/locuszoom/susie_coloc/SAIGE_merged_all_susie_coloc.pdf",height=12, width=15)
ggarrange(plot_grid(p1, cs_plt1, g, ncol = 1, rel_heights = c(2, 2, 1), align = "v"),
          plot_grid(p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, p8, cs_plt8, gg_genetracks(loc,filter_gene_name = "ALDH2"), ncol = 1, align = "v"), ncol = 2, #widths = c(2,1.5),
          labels = c("a", "b"),font.label=list(size=20))
dev.off()

```

#===================
# Pleiotropy figure
#===================
```{r}
bim <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim",data.table = F)

# final 36 variables
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HEIGHT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")
```

```{r}
run_pleiotropy <- function(trait){
  sumstats <- get_gwas_rs671(trait)
  sumstats2 <- sumstats %>%
    dplyr::mutate(sig=ifelse(neglogP>-log10(5e-8),1,0))
  return(sumstats2$sig)
}

df_pleiotropy0 <- data.frame()
for(i in c(1:length(trait_list))){
  out_plt <- run_pleiotropy(trait_list[i])
  df_pleiotropy0 <- rbind(df_pleiotropy0,out_plt)
}
df_pleiotropy0 <- as.data.frame(df_pleiotropy0)
df_pleiotropy <- df_pleiotropy0[,-c(1:200)] # to match with gene_positions in the merged version
sumstats <- get_gwas_rs671("ALCO_AMOUNT")
colnames(df_pleiotropy)<- sumstats$POS[-c(1:200)]   #position

# for(i in c(1:ncol(df_pleiotropy))){
#   colnames(df_pleiotropy)[i] <- sumstats2$BP[i]+i
# }
rownames(df_pleiotropy)<-trait_list

```

```{r}
#install.packages("gplots")
# library("gplots")
# heatmap.2(as.matrix(df_pleiotropy), scale = "col", col = bluered(100), 
#           trace = "none", density.info = "none",dendrogram = "none")

```  
```{r}
library(reshape2); library(ggplot2)
parRes_df <- melt(cbind(trait=rownames(df_pleiotropy), df_pleiotropy))
parRes_df2 <- parRes_df%>% 
      dplyr::mutate(category = case_when(
         (trait %in% c("BMI","HEIGHT","WAIST","WT")) ~ "Anthropometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardiovascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor marker",
              TRUE ~ "")) %>%
  dplyr::mutate(value1=ifelse(category=="Liver",value*2,
                       ifelse(category=="Thyroid hormone",value*3,
                       ifelse(category=="Tumor marker",value*4,
                       ifelse(category=="Kidney",value*5,
                       ifelse(category=="Hematological",value*6,
                       ifelse(category=="Cardiovascular",value*7,
                       ifelse(category=="Anthropometry",value*8,
                       ifelse(category=="Lifestyle",value*9,value))))))))) %>%
  dplyr::mutate(category = factor(category, levels=unique(category)),
                trait = factor(trait, levels=c("BMI","HEIGHT","WAIST","WT","SBP","DBP","HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS","ALP","CREAT","URIC","COFFA","SMOKA_MOD","ALCO_AMOUNT","ALB","BIL","GGT","GOT","GPT","ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN","TSH","CEA"))) %>%
  arrange(category)

cbPalette <- c("white", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

plot_pleiotropy <- ggplot(parRes_df2, 
    aes(x = variable, y = trait, fill = factor(value1))) + 
    geom_tile() + 
    #scale_fill_manual(values=colors, aesthetics = c("colour", "fill")) + 
  scale_fill_manual(values=cbPalette) +
  theme(legend.position = "none") + xlab("") + ylab("")
```

### Final
```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/locuszoom/susie_coloc/SAIGE_merged_all_susie_coloc_pleiotropy.pdf",height=12, width=15)
ggarrange(plot_grid(p1, cs_plt1, plot_pleiotropy, g, ncol = 1, rel_heights = c(2, 1, 3, 1), align = "v"),
          plot_grid(p2, cs_plt2, p3, cs_plt3, p4, cs_plt4, p5, cs_plt5, p6, cs_plt6, p7, cs_plt7, p8, cs_plt8, gg_genetracks(loc,filter_gene_name = "ALDH2"), ncol = 1, align = "v"), ncol = 2, #widths = c(2,1.5),
          labels = c("a", "b"),font.label=list(size=20))
dev.off()

```

#==========================================================
## SuSiE using sumstat: ALCO, DBP â€“ run using L=5, L=1
#==========================================================
```{r}
run_susie_L <- function(trait,L_input){

  # Merge gwas and extracting SNPs
  sumstats <- get_gwas_rs671(trait)

  # Z-score
  z_scores <- sumstats$BETA / sumstats$SE
  
  # LD matrix
  R_ref <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr12_kcps2_r.ld.gz"))
  R_ref <- as.matrix(R_ref)
  rownames(R_ref)<-colnames(R_ref)
  
  ## Run SuSiE
  fitted_rss3 <- susieR::susie_rss(z_scores, R_ref, n = sumstats$N[1], L = L_input, refine = TRUE)
  print(paste0("SuSiE converged: ",fitted_rss3$converged))
  
  #susie_plot(fitted_rss3, y="PIP")

  #ggsave(paste0("/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/plot/susie_plot_",trait,".png"), width = 15, height = 7)

  print(trait)
  return(fitted_rss3)  
}
```

## L=1
```{r}
#trait_list_L <- c("DBP","ALCO_AMOUNT")
trait_list_L <- c("ALCO_AMOUNT","GOT","GPT","GGT","SBP","DBP","COFFA","TG")
all_trait_output_L1 <- list()               # Create empty list
for(k in 1:length(trait_list_L)){
  all_trait_output_L1[[trait_list_L[k]]] <- run_susie_L(trait_list_L[k],1)
}
```

```{r}
for(k in 1:length(trait_list_L)){
  print(snp[all_trait_output_L1[[trait_list_L[k]]]$sets$cs$L1])
}
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs4646776" "rs671"    
# [1] "rs671"
# [1] "rs4646776"  "rs671"      "rs78069066"
# [1] "rs671"
```

## L=5
```{r}
trait_list_L <- c("ALCO_AMOUNT","GOT","GPT","GGT","SBP","DBP","COFFA","TG")
all_trait_output_L5 <- list()               # Create empty list
for(k in 1:length(trait_list_L)){
  all_trait_output_L5[[trait_list_L[k]]] <- run_susie_L(trait_list_L[k],5)
}
```

```{r}
for(k in 1:length(trait_list_L)){
  print(snp[all_trait_output_L5[[trait_list_L[k]]]$sets$cs$L1])
}
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs671"
# [1] "rs4646776" "rs671"    
# [1] "rs671"
```

```{r}
snp[all_trait_output$DBP$sets$cs$L1,] #rs671
snp[all_trait_output$ALCO_AMOUNT$sets$cs$L1,] #rs671
```

## Collect outputs
```{r}
trait_list <- trait_list_L
# all_trait_output <- all_trait_output_L1
all_trait_output <- all_trait_output_L5

pip_res <- c()
for(k in 1:length(trait_list)){
  bim <- fread(paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim"),data.table = F)
  snp <- bim$V2

  mean_pip0 <- mean(all_trait_output[[trait_list[k]]]$pip)
  #count_pip1 <- sum(all_trait_output[[trait_list[k]]]$pip>0)
  cs <- all_trait_output[[trait_list[k]]]$sets$cs
  N_cs <- length(cs)
  Total_size_cs <- sum(length(cs$L1),length(cs$L2),length(cs$L3),length(cs$L4),length(cs$L5),length(cs$L6),length(cs$L7),length(cs$L8),length(cs$L9),length(cs$L10), na.rm = TRUE)
  N_cs_1 <- sum(length(cs$L1)==1,length(cs$L2)==1,length(cs$L3)==1,length(cs$L4)==1,length(cs$L5)==1,length(cs$L6)==1,length(cs$L7)==1,length(cs$L8)==1,length(cs$L9)==1,length(cs$L10)==1, na.rm = TRUE)
  size_cs <- paste(c(length(cs$L1),length(cs$L2),length(cs$L3),length(cs$L4),length(cs$L5),length(cs$L6),length(cs$L7),length(cs$L8),length(cs$L9),length(cs$L10)),sep=",",collapse = ",")

  SNP <- paste(snp[c(cs$L1,cs$L2,cs$L3,cs$L4,cs$L5,cs$L6,cs$L7,cs$L8,cs$L9,cs$L10)],sep=",",collapse = ",")
  Is_rs671 <- str_detect(SNP, "rs671")
  snp_index <- which(snp=="rs671")
  PIP_rs671 <- all_trait_output[[trait_list[k]]]$pip[snp_index] #rs671: 622
 
  #mean_pip1 <- cbind(trait_list[k], mean_pip0, count_pip1, nrow(snp), N_cs, size_cs, SNP)
  mean_pip1 <- cbind(trait_list[k], mean_pip0, length(snp), N_cs, N_cs_1, Total_size_cs, size_cs, SNP, Is_rs671, PIP_rs671)
  
  pip_res <- rbind(pip_res, mean_pip1)
}
pip_res <- as.data.frame(pip_res)
colnames(pip_res) <- c("trait","mean_PIP", "Number of SNPs fine-mapped","Number of CS", "Number of CS with 1 variant","Total size of CS (#variants)","Size of each CS","SNPs in cs","Is_rs671","PIP_rs671")

```
```{r}
#write.csv(pip_res,"/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_mean_pip_rs671_L1.csv",row.names = FALSE)
write.csv(pip_res,"/n/holylfs05/LABS/kraft_lab/Lab/yjee/finemap/SAIGE_mean_pip_rs671_L5.csv",row.names = FALSE)

```

#==========================================================
# 09.29.2024
# conditional analysis
#==========================================================
```{r}
gene_of_interest = "ALDH2"
get_input_locuszoom_cond <- function(sumstats_path_filename=NULL,
                                trait=NULL, chr=NULL,
                         bim_path_filename, 
                         ldm_path_filename, 
                         snp_of_interest=NULL, #default is most significant SNP
                         gene_of_interest){
  # 1. Load sumstat & bim file used to create LD matrix
  #sumstats_path_filename <- "~/GLP1R.csv"
  if(!is.null(sumstats_path_filename)){
    sumstats <- fread(sumstats_path_filename, data.table = F)
    sumstats$Z <- sumstats$bC/sumstats$bC_se
    } else {
      gwas_fn = paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/gwas/SAIGE/loco/")
      sumstats <- fread(paste0(gwas_fn,"SAIGE_",trait,"_INFO.txt.gz"), data.table = F) 
      }

  #bim_path_filename <- "/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_GLP1R.bim"
  bim0 <- fread(bim_path_filename) 
  #snp <- bim$V2
  #sum(sumstats$SNP!=bim0$V2)
  
  #3. Match SNP list between sumstats and LD matrix
  sumstats02 <- sumstats %>%
    dplyr::rename(CHR = Chr, POS=bp) %>% #names(sumstats)[grepl("CHR",names(sumstats))]
    dplyr::mutate(neglogP = -log10(pC)) %>%
    dplyr::select(CHR, POS, SNP, neglogP)
  
  # re-Filter conditional analysis SNPs (except rs671)
  bim <- bim0 %>% dplyr::filter(V2%in%sumstats$SNP)
  snp <- bim$V2
  
  sumstats2 <- left_join(bim,sumstats02,by=c("V2"="SNP")) %>%
    dplyr::select(CHR, POS, V2, neglogP) %>%
    dplyr::rename(SNP=V2)
  
  # 2. Load LD matrix
  #ldm_path_filename <- "/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_r2_GLP1R.ld.gz"
  R_ref <- fread(ldm_path_filename)
  R_ref <- R_ref[-622,-622] #which(bim0$V2=="rs671")
  R_ref <- as.matrix(R_ref)
  colnames(R_ref) <- snp
  rownames(R_ref) <- snp
  
  # convert LD matrix to long format
  R_ref_ld <- as.data.frame(as.table(cor(R_ref))) 
  R_ref_ld <- as.data.frame(cbind(R_ref_ld,rep(sumstats2$CHR[1],nrow(R_ref)*nrow(R_ref)),rep(sumstats2$CHR[1],nrow(R_ref)*nrow(R_ref)),rep(sumstats2$POS,each=nrow(R_ref)),rep(sumstats2$POS,each=nrow(R_ref))))
  R_ref_ld$R2<-R_ref_ld$Freq^2
  R_ref_ld <- R_ref_ld[,-3]
  names(R_ref_ld)<-c("SNP_A","SNP_B","CHR_A","CHR_B","POS_A","POS_B","R2")
  # CHR_A         POS_A         SNP_A  CHR_B         POS_B         SNP_B           R2 
  
  # extract the most significant SNP from LD long format
  if(is.null(snp_of_interest)){
    most_sig_snp <- sumstats2$SNP[order(-sumstats2$neglogP)][1]
  } else most_sig_snp <- snp_of_interest
  
  gene <- gene_of_interest
  
  R_ref_ld <- R_ref_ld[R_ref_ld$SNP_A==most_sig_snp,]
  
  R_ref_ld2 <- R_ref_ld %>%
    dplyr::rename(SNP=SNP_B, r2=R2)

  gwas_r2 <- left_join(sumstats2, R_ref_ld2, by="SNP") %>%
    dplyr::select(CHR, POS, SNP, neglogP, r2) %>%
    dplyr::rename(rsid=SNP)

  return(list(sumstats2, R_ref_ld, most_sig_snp, gene,
              gwas_r2))
}

```

```{r}
get_loc_locuszoomr_cond <- function(get_input_list, window){
  
  if(!is.null(get_input_list$sumstats_path_filename)){
    sumstats_path_filename = get_input_list$sumstats_path_filename
    #trait = NULL; chr = NULL
    } else {
      sumstats_path_filename = NULL
      trait = get_input_list$trait; chr = get_input_list$chr
    }
  bim_path_filename = get_input_list$bim_path_filename
  ldm_path_filename = get_input_list$ldm_path_filename
  snp_of_interest = get_input_list$snp_of_interest
  gene_of_interest = get_input_list$gene_of_interest

  res <- get_input_locuszoom_cond(sumstats_path_filename = sumstats_path_filename,
                             trait = trait, chr = chr,
                             bim_path_filename = bim_path_filename,
                             ldm_path_filename = ldm_path_filename,
                             snp_of_interest = snp_of_interest,
                             gene_of_interest = gene_of_interest)
  
  gwas_r2 <- res[[5]]
  most_sig_snp <- res[[3]]
  gene <- res[[4]]
  
  if (require(EnsDb.Hsapiens.v75)) {
    loc <- locus(gwas_r2, gene = gene, flank = window, #1e5
             LD = "r2", ens_db = "EnsDb.Hsapiens.v75", index_snp = most_sig_snp, yvar = "neglogP")
    locus_plot(loc, labels = "index", label_x = c(4, -5), ylab="-log10(P-value)")
  }
  
  return(loc)
}
```

Format input_list 
```{r}
GWAS_input_list_ft_cond <- function(gene,trait,chr,eQTL_snp){
  input_list2 <- list(trait=trait,chr=chr,
                      sumstats_path_filename=paste0("/n/holyscratch01/kraft_lab/yjee/GCTA/COJO_output/",trait,"/",trait,"_rs671.cma.cojo"), 
                      bim_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr",chr,".bim"), 
                     ldm_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr",chr,"_kcps2_r.ld.gz"),
                       snp_of_interest=eQTL_snp, 
                    gene_of_interest=gene)
  return(input_list2)
}
```

```{r}
gene="ALDH2"
trait1="ALCO_AMOUNT"
trait2="GOT"
trait3="GPT"
trait4="GGT"
trait5="SBP"
trait6="DBP"
trait7="COFFA"
trait8="TG"

chr=12
eQTL_snp="rs671"
eQTL_snp=NULL
kcps2_gwas_input_list1 <- GWAS_input_list_ft_cond(gene,trait1,chr,eQTL_snp)
loc <- get_loc_locuszoomr_cond(get_input_list = kcps2_gwas_input_list1, window=5e5)
```

```{r}
grob1 <- grobTree(textGrob("Alcohol amount", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob(loc$index_snp, x=0.5, y=0.98, hjust=0,
  gp=gpar(col="black", fontsize=10)))

p1 <- gg_scatter(loc, xticks = FALSE, legend_pos = "topright", border = TRUE, ylab="-log10(P-value)") + annotation_custom(grob1) 

g <- gg_genetracks(loc,maxrows = 3)
```

# Pleiotropy figure
## Get GWAS
```{r}
trait="ALCO_AMOUNT"
get_gwas_rs671_cond=function(trait){
  
  sumstats <- fread(paste0("/n/holyscratch01/kraft_lab/yjee/GCTA/COJO_output/",trait,"/",trait,"_rs671.cma.cojo"), data.table = F)

  bim_path_filename = paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr",chr,".bim")
  bim0 <- fread(bim_path_filename) 
  #snp <- bim$V2
  #sum(sumstats$SNP!=bim0$V2)
  
  #3. Match SNP list between sumstats and LD matrix
  sumstats02 <- sumstats %>%
    dplyr::rename(CHR = Chr, POS=bp) %>% #names(sumstats)[grepl("CHR",names(sumstats))]
    dplyr::mutate(neglogP = -log10(pC)) %>%
    dplyr::select(CHR, POS, SNP, neglogP)
  
  # re-Filter conditional analysis SNPs (except rs671)
  bim <- bim0 %>% dplyr::filter(V2%in%sumstats$SNP)
  snp <- bim$V2
  
  sumstats2 <- left_join(bim,sumstats02,by=c("V2"="SNP")) %>%
    dplyr::select(CHR, POS, V2, neglogP) %>%
    dplyr::rename(SNP=V2)
  
  print(trait)
  return(sumstats2)
}
```

```{r}
# bim <- fread("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr12.bim",data.table = F)

# final 36 variables
trait_list=c("LDL","TG","HDL","CHO","FBS","INSULIN","GOT","GPT","GGT","BIL","ALB","TSH","CEA",
             "CREAT","ADIPO","URIC","ALP","HB","HCT","MCH","MCHC","MCV","RBC","RDW","WBC","PLT","EOS",
             "SBP","DBP","WT","HEIGHT","BMI","WAIST","SMOKA_MOD","ALCO_AMOUNT","COFFA")
```

```{r}
run_pleiotropy <- function(trait){
  sumstats <- get_gwas_rs671_cond(trait)
  sumstats2 <- sumstats %>%
    dplyr::mutate(sig=ifelse(neglogP>-log10(5e-8),1,0))
  return(sumstats2$sig)
}

df_pleiotropy0 <- data.frame()
for(i in c(1:length(trait_list))){
  out_plt <- run_pleiotropy(trait_list[i])
  df_pleiotropy0 <- rbind(df_pleiotropy0,out_plt)
}
df_pleiotropy0 <- as.data.frame(df_pleiotropy0)
df_pleiotropy <- df_pleiotropy0[,-c(1:200)] # to match with gene_positions in the merged version
sumstats <- get_gwas_rs671_cond("ALCO_AMOUNT")
colnames(df_pleiotropy)<- sumstats$POS[-c(1:200)]   #position

# for(i in c(1:ncol(df_pleiotropy))){
#   colnames(df_pleiotropy)[i] <- sumstats2$BP[i]+i
# }
rownames(df_pleiotropy)<-trait_list

```

```{r}
#install.packages("gplots")
# library("gplots")
# heatmap.2(as.matrix(df_pleiotropy), scale = "col", col = bluered(100), 
#           trace = "none", density.info = "none",dendrogram = "none")

```  
```{r}
library(reshape2); library(ggplot2)
parRes_df <- melt(cbind(trait=rownames(df_pleiotropy), df_pleiotropy))
parRes_df2 <- parRes_df%>% 
      dplyr::mutate(category = case_when(
         (trait %in% c("BMI","HEIGHT","WAIST","WT")) ~ "Anthropometry",
                                         (trait %in% c("SBP","DBP")) ~ "Cardiovascular",
                                   (trait %in% c("HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS")) ~ "Hematological",
                                   (trait %in% c("ALP","CREAT","URIC")) ~ "Kidney",
                                   (trait %in% c("COFFA","SMOKA_MOD","ALCO_AMOUNT")) ~ "Lifestyle",
                                   (trait %in% c("ALB","BIL","GGT","GOT","GPT")) ~ "Liver",
              (trait %in% c("ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN")) ~ "Metabolic",
              (trait %in% c("TSH")) ~ "Thyroid hormone",
              (trait %in% c("CEA")) ~ "Tumor marker",
              TRUE ~ "")) %>%
  dplyr::mutate(value1=ifelse(category=="Liver",value*2,
                       ifelse(category=="Thyroid hormone",value*3,
                       ifelse(category=="Tumor marker",value*4,
                       ifelse(category=="Kidney",value*5,
                       ifelse(category=="Hematological",value*6,
                       ifelse(category=="Cardiovascular",value*7,
                       ifelse(category=="Anthropometry",value*8,
                       ifelse(category=="Lifestyle",value*9,value))))))))) %>%
  dplyr::mutate(category = factor(category, levels=unique(category)),
                trait = factor(trait, levels=c("BMI","HEIGHT","WAIST","WT","SBP","DBP","HB","HCT","MCHC","MCV","RBC","RDW","MCH","WBC","PLT","EOS","ALP","CREAT","URIC","COFFA","SMOKA_MOD","ALCO_AMOUNT","ALB","BIL","GGT","GOT","GPT","ADIPO","CHO","FBS","HDL","LDL","TG","INSULIN","TSH","CEA"))) %>%
  arrange(category)

cbPalette <- c("white", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

plot_pleiotropy <- ggplot(parRes_df2, 
    aes(x = variable, y = trait, fill = factor(value1))) + 
    geom_tile() + 
    #scale_fill_manual(values=colors, aesthetics = c("colour", "fill")) + 
  scale_fill_manual(values=cbPalette) +
  theme(legend.position = "none") + xlab("") + ylab("")
```

```{r}
pdf("/n/holylfs05/LABS/kraft_lab/Lab/yjee/locuszoom/susie_coloc/SAIGE_merged_all_susie_coloc_pleiotropy_cond.pdf",height=12, width=7)
ggarrange(plot_grid(p1, plot_pleiotropy, g, ncol = 1, rel_heights = c(2, 3, 1), align = "v"),ncol = 1)
dev.off()

```




# Old script
```{r}
cs_plt <- data.frame()
  for(i in c(1:length(susie_output$sets$cs))){
    cs <- rep(susie_output$sets$cs_index[[i]], length(susie_output$sets$cs[[i]]))
    pip <- susie_output$pip[susie_output$sets$cs[[i]]]
    s <- snp[susie_output$sets$cs[[i]]]
    p <- pos[susie_output$sets$cs[[i]]]
    out <- cbind(s,p,pip,cs)
    cs_plt <- rbind(cs_plt, out)
  }
  names(cs_plt) <- c("SNP","POS","PIP","CS")
  cs_plt <- cs_plt %>%
    mutate(CS = factor(CS,levels=c("1","2","3","4","5","6","7","8","9","10")))

cs_plt1 <- cs_plt %>%
  ggplot(aes(as.numeric(POS),as.numeric(PIP), col=factor(CS)))+ 
  #geom_point(position = position_dodge(0.5), size = 5)+
  geom_point(
    label=cs_plt$SNP, size=2,
    nudge_x = 0.01,
    check_overlap = T) +
  # geom_label_repel(
  #   aes(label=trait), size=4, box.padding = unit(0.5, "lines")
  # )+
  ggpubr::theme_pubr(border = TRUE) +
  geom_abline(intercept = 0, slope = 1, linetype="dotted") + 
  #xlim(0,1) + 
  ylim(0,1) + xlab("") + ylab("PIP") + 
  scale_y_continuous(breaks=seq(0,1,0.5)) +
  theme(legend.position = c(0.9,0.6), axis.text.x=element_blank(), 
      axis.ticks.x=element_blank()) + 
  guides(col = guide_legend(ncol = 2, title="95% CS"))
```


Format input_list 
```{r}
GWAS_input_list_ft <- function(gene,trait,chr,eQTL_snp){
  input_list2 <- list(trait=trait,chr=chr,
                      bim_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/kcps2_ld_chr",chr,".bim"), 
                     ldm_path_filename=paste0("/n/holylfs05/LABS/kraft_lab/Lab/KCPS2/ldm/extract/chr",chr,"_kcps2_r.ld.gz"),
                       snp_of_interest=eQTL_snp, 
                    gene_of_interest=gene)
  return(input_list2)
}
```

```{r}
grob1 <- grobTree(textGrob("Alcohol amount", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob(loc$index_snp, x=0.5, y=0.98, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob2 <- grobTree(textGrob("GOT", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob3 <- grobTree(textGrob("GPT", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob4 <- grobTree(textGrob("GGT", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob5 <- grobTree(textGrob("SBP", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob6 <- grobTree(textGrob("DBP", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob7 <- grobTree(textGrob("Coffee intake", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
grob8 <- grobTree(textGrob("TG", x=0.01,  y=0.90, hjust=0,
  gp=gpar(col="black", fontsize=13, fontface="bold")),
  textGrob("coloc.pp4=1", x=0.01,  y=0.60, hjust=0,
  gp=gpar(col="black", fontsize=10)))
```
